// @maptiler/sdk/dist/maptiler-sdk.mjs
var rt = Object.defineProperty;
var me = (r) => {
  throw TypeError(r);
};
var at = (r, e, t) => e in r ? rt(r, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : r[e] = t;
var T = (r, e, t) => at(r, typeof e != "symbol" ? e + "" : e, t), re = (r, e, t) => e.has(r) || me("Cannot " + t);
var L = (r, e, t) => (re(r, e, "read from private field"), t ? t.call(r) : e.get(r)), U = (r, e, t) => e.has(r) ? me("Cannot add the same private member more than once") : e instanceof WeakSet ? e.add(r) : e.set(r, t), $ = (r, e, t, o) => (re(r, e, "write to private field"), o ? o.call(r, t) : e.set(r, t), t), k = (r, e, t) => (re(r, e, "access private method"), t);
import f from "maplibre-gl";
import { Language as nt, getLanguageInfoFromCode as fe, config as ae, MapStyle as lt, mapStylePresetList as it, expandMapStyle as st, MapStyleVariant as ct, ReferenceMapStyle as ut, getLanguageInfoFromFlag as dt, toLanguageInfo as pt, geolocation as ge } from "@maptiler/client";
import { MapStyle as oa, MapStyleVariant as ra, ReferenceMapStyle as aa, ServiceError as na, areSameLanguages as la, bufferToPixelDataBrowser as ia, circumferenceAtLatitude as sa, coordinates as ca, data as ua, elevation as da, expandMapStyle as pa, geocoding as ha, geolocation as ma, getAutoLanguage as fa, getBufferToPixelDataParser as ga, getLanguageInfoFromCode as ya, getLanguageInfoFromFlag as va, getLanguageInfoFromKey as La, getTileCache as wa, isLanguageInfo as Ca, mapStylePresetList as ba, math as Sa, misc as Aa, staticMaps as Ta, styleToStyle as Ra, toLanguageInfo as _a } from "@maptiler/client";
import ht from "events";
import { v4 as mt } from "uuid";
import { Base64 as ft } from "js-base64";
const gt = "@maptiler/sdk", yt = "2.3.0", vt = "The Javascript & TypeScript map SDK tailored for MapTiler Cloud", Lt = "dist/maptiler-sdk.mjs", wt = "dist/maptiler-sdk.d.ts", Ct = "dist/maptiler-sdk.css", bt = "module", St = {
  ".": {
    import: "./dist/maptiler-sdk.mjs",
    types: "./dist/maptiler-sdk.d.ts"
  },
  "./dist/maptiler-sdk.css": {
    import: "./dist/maptiler-sdk.css"
  },
  "./style.css": {
    import: "./dist/maptiler-sdk.css"
  }
}, At = [
  "maptiler",
  "map",
  "sdk",
  "webmap",
  "cloud",
  "webGL",
  "maplibre"
], Tt = "https://docs.maptiler.com/sdk-js/", Rt = "BSD-3-Clause", _t = {
  type: "git",
  url: "https://github.com/maptiler/maptiler-sdk-js.git"
}, Et = {
  biome: "biome check --max-diagnostics=1000",
  "biome:fix": "npx @biomejs/biome check --max-diagnostics=1000 --write",
  doc: "rm -rf docs/* && typedoc --out docs && cp -r images docs/",
  test: "vitest run",
  "build-css": "node scripts/replace-path-with-content.js src/style/style_template.css dist/tmp_maptiler-sdk.css && cat node_modules/maplibre-gl/dist/maplibre-gl.css dist/tmp_maptiler-sdk.css > dist/maptiler-sdk.css && rm dist/tmp_maptiler-sdk.css && cp dist/maptiler-sdk.css build/maptiler-sdk.css",
  "build-umd": "tsc && NODE_ENV=production vite build -c vite.config-umd.ts",
  "build-es": "tsc && NODE_ENV=production vite build -c vite.config-es.ts",
  build: "npm run build-es; npm run build-umd; npm run build-css",
  make: "npm run biome:fix && npm run build",
  "dev-umd": "npm run build-css && tsc && NODE_ENV=dev vite build -w -c vite.config-umd.ts",
  help: "vite build --help"
}, xt = "MapTiler", It = {
  "@biomejs/biome": "1.8.3",
  "@types/uuid": "^9.0.2",
  "@types/xmldom": "^0.1.31",
  "@xmldom/xmldom": "^0.8.10",
  serve: "^14.2.0",
  terser: "^5.17.1",
  typedoc: "^0.24.8",
  typescript: "^5.1.6",
  vite: "^5.3.4",
  "vite-plugin-dts": "^4.0.0-beta.0",
  vitest: "^0.34.2"
}, Mt = {
  "@maptiler/client": "^2.0.0",
  events: "^3.3.0",
  "js-base64": "^3.7.4",
  "maplibre-gl": "^4.7.0",
  uuid: "^9.0.0"
}, Pt = {
  name: gt,
  version: yt,
  description: vt,
  module: Lt,
  types: wt,
  style: Ct,
  type: bt,
  exports: St,
  keywords: At,
  homepage: Tt,
  license: Rt,
  repository: _t,
  scripts: Et,
  author: xt,
  devDependencies: It,
  dependencies: Mt
}, C = {
  /**
   * Language mode to display labels in both the local language and the language of the visitor's device, concatenated.
   * Note that if those two languages are the same, labels won't be duplicated.
   */
  VISITOR: {
    code: null,
    flag: "visitor",
    name: "Visitor",
    latin: !0,
    isMode: !0,
    geocoding: !1
  },
  /**
   * Language mode to display labels in both the local language and English, concatenated.
   * Note that if those two languages are the same, labels won't be duplicated.
   */
  VISITOR_ENGLISH: {
    code: null,
    flag: "visitor_en",
    name: "Visitor English",
    latin: !0,
    isMode: !0,
    geocoding: !1
  },
  /**
   * Language mode to display labels in a language enforced in the style.
   */
  STYLE: { code: null, flag: "style", name: "Style", latin: !1, isMode: !0, geocoding: !1 },
  /**
   * Language mode to display labels in a language enforced in the style. The language cannot be further modified.
   */
  STYLE_LOCK: {
    code: null,
    flag: "style_lock",
    name: "Style Lock",
    latin: !1,
    isMode: !0,
    geocoding: !1
  },
  ...nt
};
function ye() {
  if (typeof navigator > "u") {
    const e = Intl.DateTimeFormat().resolvedOptions().locale.split("-")[0], t = fe(e);
    return t || C.ENGLISH;
  }
  return Array.from(new Set(navigator.languages.map((e) => e.split("-")[0]))).map((e) => fe(e)).filter((e) => e)[0] ?? C.LOCAL;
}
const O = {
  maptilerLogoURL: "https://api.maptiler.com/resources/logo.svg",
  maptilerURL: "https://www.maptiler.com/",
  maptilerApiHost: "api.maptiler.com",
  rtlPluginURL: "https://cdn.maptiler.com/mapbox-gl-rtl-text/v0.2.3/mapbox-gl-rtl-text.min.js",
  primaryLanguage: C.STYLE,
  secondaryLanguage: C.LOCAL,
  terrainSourceURL: "https://api.maptiler.com/tiles/terrain-rgb-v2/tiles.json",
  terrainSourceId: "maptiler-terrain"
};
Object.freeze(O);
const Be = mt();
class kt extends ht {
  constructor() {
    super(...arguments);
    /**
     * The primary language. By default, the language of the web browser is used.
     */
    T(this, "primaryLanguage", O.primaryLanguage);
    /**
     * The secondary language, to overwrite the default language defined in the map style.
     * This settings is highly dependant on the style compatibility and may not work in most cases.
     */
    T(this, "secondaryLanguage");
    /**
     * Setting on whether of not the SDK runs with a session logic.
     * A "session" is started at the initialization of the SDK and finished when the browser
     * page is being refreshed.
     * When `session` is enabled (default: true), the extra URL param `mtsid` is added to queries
     * on the MapTiler Cloud API. This allows MapTiler to enable "session based billing".
     */
    T(this, "session", !0);
    /**
     * Enables client-side caching of requests for tiles and fonts.
     * The cached requests persist multiple browser sessions and will be reused when possible.
     * Works only for requests to the MapTiler Cloud API when sessions are enabled.
     */
    T(this, "caching", !0);
    /**
     * Unit to be used
     */
    T(this, "_unit", "metric");
    /**
     * MapTiler Cloud API key
     */
    T(this, "_apiKey", "");
  }
  /**
   * Set the unit system
   */
  set unit(t) {
    this._unit = t, this.emit("unit", t);
  }
  /**
   * Get the unit system
   */
  get unit() {
    return this._unit;
  }
  /**
   * Set the MapTiler Cloud API key
   */
  set apiKey(t) {
    this._apiKey = t, ae.apiKey = t, this.emit("apiKey", t);
  }
  /**
   * Get the MapTiler Cloud API key
   */
  get apiKey() {
    return this._apiKey;
  }
  /**
   * Set a the custom fetch function to replace the default one
   */
  set fetch(t) {
    ae.fetch = t;
  }
  /**
   * Get the fetch fucntion
   */
  get fetch() {
    return ae.fetch;
  }
}
const M = new kt(), { addProtocol: ve } = f, ie = "localcache_source", se = "localcache", Ft = "maptiler_sdk", Ot = 1e3, Bt = 100, ce = typeof caches < "u";
function Nt(r, e) {
  if (ce && M.caching && M.session && r.host === O.maptilerApiHost) {
    if (e === "Source" && r.href.includes("tiles.json"))
      return r.href.replace("https://", `${ie}://`);
    if (e === "Tile" || e === "Glyphs")
      return r.href.replace("https://", `${se}://`);
  }
  return r.href;
}
let ne;
async function Ne() {
  return ne || (ne = await caches.open(Ft)), ne;
}
let Le = 0;
async function zt() {
  const r = await Ne(), e = await r.keys(), t = e.slice(0, Math.max(e.length - Ot, 0));
  for (const o of t)
    r.delete(o);
}
function Dt() {
  ve(
    ie,
    async (r, e) => {
      if (!r.url) throw new Error("");
      r.url = r.url.replace(`${ie}://`, "https://");
      const t = r;
      t.signal = e.signal;
      const o = await fetch(r.url, t), a = await o.json();
      return a.tiles && a.tiles.length > 0 && (a.tiles[0] += `&last-modified=${o.headers.get("Last-Modified")}`), {
        data: a,
        cacheControl: o.headers.get("Cache-Control"),
        expires: o.headers.get("Expires")
      };
    }
  ), ve(
    se,
    async (r, e) => {
      if (!r.url) throw new Error("");
      r.url = r.url.replace(`${se}://`, "https://");
      const t = new URL(r.url), o = new URL(t);
      o.searchParams.delete("mtsid"), o.searchParams.delete("key");
      const a = o.toString(), n = new URL(t);
      n.searchParams.delete("last-modified");
      const l = n.toString(), i = async (d) => ({
        data: await d.arrayBuffer(),
        cacheControl: d.headers.get("Cache-Control"),
        expires: d.headers.get("Expires")
      }), u = await Ne(), h = await u.match(a);
      if (h)
        return i(h);
      const c = r;
      c.signal = e.signal;
      const s = await fetch(l, c);
      return s.status >= 200 && s.status < 300 && (u.put(a, s.clone()).catch(() => {
      }), ++Le > Bt && (zt(), Le = 0)), i(s);
    }
  );
}
function Ut() {
  if (typeof window > "u") return;
  const r = f.getRTLTextPluginStatus();
  if (r === "unavailable" || r === "requested")
    try {
      f.setRTLTextPlugin(O.rtlPluginURL, !0);
    } catch {
    }
}
function Gt(r, e) {
  for (const t of r)
    typeof e[t] == "function" && (e[t] = e[t].bind(e));
}
function Z(r, e, t) {
  const o = window.document.createElement(r);
  return e !== void 0 && (o.className = e), t && t.appendChild(o), o;
}
function ze(r) {
  r.parentNode && r.parentNode.removeChild(r);
}
function we(r, e) {
  let t = null;
  try {
    t = new URL(r);
  } catch {
    return {
      url: r
    };
  }
  return t.host === O.maptilerApiHost && (t.searchParams.has("key") || t.searchParams.append("key", M.apiKey), M.session && t.searchParams.append("mtsid", Be)), {
    url: Nt(t, e)
  };
}
function Ce(r) {
  return (e, t) => {
    if (r != null) {
      const o = r(e, t), a = we((o == null ? void 0 : o.url) ?? "", t);
      return {
        ...o,
        ...a
      };
    }
    return we(e, t);
  };
}
function De() {
  return Math.random().toString(36).substring(2);
}
function ee(r) {
  return /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/gi.test(r);
}
function Ht(r) {
  try {
    return JSON.parse(r);
  } catch {
  }
  return null;
}
function Vt() {
  return document.createElement("canvas").getContext("webgl2") ? null : typeof WebGL2RenderingContext < "u" ? "Graphic rendering with WebGL2 has been disabled or is not supported by your graphic card. The map cannot be displayed." : "Your browser does not support graphic rendering with WebGL2. The map cannot be displayed.";
}
function $t(r) {
  const e = Vt();
  if (!e) return;
  let t = null;
  if (typeof r == "string" ? t = document.getElementById(r) : r instanceof HTMLElement && (t = r), !t)
    throw new Error("The Map container must be provided.");
  const o = document.createElement("div");
  throw o.innerHTML = e, o.classList.add("no-webgl-support-div"), t.appendChild(o), new Error(e);
}
let xo = class extends f.Marker {
  addTo(e) {
    return super.addTo(e);
  }
};
class Mo extends f.Popup {
  addTo(e) {
    return super.addTo(e);
  }
}
class Po extends f.Style {
  constructor(e, t = {}) {
    super(e, t);
  }
}
class ko extends f.CanvasSource {
  onAdd(e) {
    super.onAdd(e);
  }
}
class Fo extends f.GeoJSONSource {
  onAdd(e) {
    super.onAdd(e);
  }
}
class Oo extends f.ImageSource {
  onAdd(e) {
    super.onAdd(e);
  }
}
class Bo extends f.RasterTileSource {
  onAdd(e) {
    super.onAdd(e);
  }
}
class No extends f.RasterDEMTileSource {
  onAdd(e) {
    super.onAdd(e);
  }
}
class zo extends f.VectorTileSource {
  onAdd(e) {
    super.onAdd(e);
  }
}
class Do extends f.VideoSource {
  onAdd(e) {
    super.onAdd(e);
  }
}
class jt extends f.NavigationControl {
  onAdd(e) {
    return super.onAdd(e);
  }
}
class Kt extends f.GeolocateControl {
  onAdd(e) {
    return super.onAdd(e);
  }
}
class Uo extends f.AttributionControl {
  onAdd(e) {
    return super.onAdd(e);
  }
}
class Zt extends f.LogoControl {
  onAdd(e) {
    return super.onAdd(e);
  }
}
class Wt extends f.ScaleControl {
  onAdd(e) {
    return super.onAdd(e);
  }
}
class Yt extends f.FullscreenControl {
  onAdd(e) {
    return super.onAdd(e);
  }
}
class Go extends f.TerrainControl {
  onAdd(e) {
    return super.onAdd(e);
  }
}
class Ho extends f.BoxZoomHandler {
  constructor(e, t) {
    super(e, t);
  }
}
class Vo extends f.ScrollZoomHandler {
  constructor(e, t) {
    super(e, t);
  }
}
class $o extends f.CooperativeGesturesHandler {
  constructor(e, t) {
    super(e, t);
  }
}
class jo extends f.KeyboardHandler {
  constructor(e) {
    super(e);
  }
}
class Ko extends f.TwoFingersTouchPitchHandler {
  constructor(e) {
    super(e);
  }
}
class Zo extends f.MapWheelEvent {
  constructor(e, t, o) {
    super(e, t, o);
  }
}
class Wo extends f.MapTouchEvent {
  constructor(e, t, o) {
    super(e, t, o);
  }
}
class Yo extends f.MapMouseEvent {
  constructor(e, t, o, a = {}) {
    super(e, t, o, a);
  }
}
class be extends Zt {
  constructor(t = {}) {
    super(t);
    T(this, "logoURL", "");
    T(this, "linkURL", "");
    this.logoURL = t.logoURL ?? O.maptilerLogoURL, this.linkURL = t.linkURL ?? O.maptilerURL;
  }
  onAdd(t) {
    this._map = t, this._compact = this.options.compact ?? !1, this._container = window.document.createElement("div"), this._container.className = "maplibregl-ctrl";
    const o = window.document.createElement("a");
    return o.style.backgroundRepeat = "no-repeat", o.style.cursor = "pointer", o.style.display = "block", o.style.height = "23px", o.style.margin = "0 0 -4px -4px", o.style.overflow = "hidden", o.style.width = "88px", o.style.backgroundImage = `url(${this.logoURL})`, o.style.backgroundSize = "100px 30px", o.style.width = "100px", o.style.height = "30px", o.target = "_blank", o.rel = "noopener", o.href = this.linkURL, o.setAttribute("aria-label", "MapTiler logo"), o.setAttribute("rel", "noopener"), this._container.appendChild(o), this._container.style.display = "block", this._map.on("resize", this._updateCompact), this._updateCompact(), this._container;
  }
}
function Se(r) {
  return r ? typeof r == "string" || r instanceof String ? !r.startsWith("http") && r.toLowerCase().includes(".json") ? r : st(r) : r instanceof ct ? r.getExpandedStyleURL() : r instanceof ut ? r.getDefaultVariant().getExpandedStyleURL() : r : lt[it[0].referenceStyleID].getDefaultVariant().getExpandedStyleURL();
}
class qt {
  constructor() {
    T(this, "_map");
    T(this, "_container");
    T(this, "_terrainButton");
    Gt(["_toggleTerrain", "_updateTerrainIcon"], this);
  }
  onAdd(e) {
    return this._map = e, this._container = Z("div", "maplibregl-ctrl maplibregl-ctrl-group"), this._terrainButton = Z("button", "maplibregl-ctrl-terrain", this._container), Z("span", "maplibregl-ctrl-icon", this._terrainButton).setAttribute("aria-hidden", "true"), this._terrainButton.type = "button", this._terrainButton.addEventListener("click", this._toggleTerrain), this._updateTerrainIcon(), this._map.on("terrain", this._updateTerrainIcon), this._container;
  }
  onRemove() {
    ze(this._container), this._map.off("terrain", this._updateTerrainIcon), this._map = void 0;
  }
  _toggleTerrain() {
    this._map.hasTerrain() ? this._map.disableTerrain() : this._map.enableTerrain(), this._updateTerrainIcon();
  }
  _updateTerrainIcon() {
    this._terrainButton.classList.remove("maplibregl-ctrl-terrain"), this._terrainButton.classList.remove("maplibregl-ctrl-terrain-enabled"), this._map.hasTerrain() ? (this._terrainButton.classList.add("maplibregl-ctrl-terrain-enabled"), this._terrainButton.title = this._map._getUIString("TerrainControl.Disable")) : (this._terrainButton.classList.add("maplibregl-ctrl-terrain"), this._terrainButton.title = this._map._getUIString("TerrainControl.Enable"));
  }
}
class Jt extends jt {
  constructor(t = {}) {
    super({
      showCompass: t.showCompass ?? !0,
      showZoom: t.showZoom ?? !0,
      visualizePitch: t.visualizePitch ?? !0
    });
    /**
     * Overloading: Limit how flat the compass icon can get
     */
    T(this, "_rotateCompassArrow", () => {
      const t = this.options.visualizePitch ? `scale(${Math.min(
        1.5,
        1 / Math.cos(this._map.transform.pitch * (Math.PI / 180)) ** 0.5
      )}) rotateX(${Math.min(70, this._map.transform.pitch)}deg) rotateZ(${this._map.transform.angle * (180 / Math.PI)}deg)` : `rotate(${this._map.transform.angle * (180 / Math.PI)}deg)`;
      this._compassIcon.style.transform = t;
    });
    this._compass && (this._compass.removeEventListener("click", this._compass.clickFunction), this._compass.addEventListener("click", (o) => {
      this._map.getPitch() === 0 ? this._map.easeTo({ pitch: Math.min(this._map.getMaxPitch(), 80) }) : this.options.visualizePitch ? this._map.resetNorthPitch({}, { originalEvent: o }) : this._map.resetNorth({}, { originalEvent: o });
    }));
  }
  /**
   * Overloading: the button now stores its click callback so that we can later on delete it and replace it
   */
  _createButton(t, o) {
    const a = super._createButton(t, o);
    return a.clickFunction = o, a;
  }
}
const Ae = f.Marker, Te = f.LngLat, Xt = f.LngLatBounds;
class Qt extends Kt {
  constructor() {
    super(...arguments);
    T(this, "lastUpdatedCenter", new Te(0, 0));
    /**
     * Update the camera location to center on the current position
     *
     * @param {Position} position the Geolocation API Position
     * @private
     */
    T(this, "_updateCamera", (t) => {
      var c, s;
      const o = new Te(t.coords.longitude, t.coords.latitude), a = t.coords.accuracy, l = {
        bearing: this._map.getBearing(),
        ...this.options.fitBoundsOptions,
        linear: !0
      }, i = this._map.getZoom();
      i > (((s = (c = this.options) == null ? void 0 : c.fitBoundsOptions) == null ? void 0 : s.maxZoom) ?? 30) && (l.zoom = i), this._map.fitBounds(Xt.fromLngLat(o, a), l, {
        geolocateSource: !0
        // tag this camera change so it won't cause the control to change to background state
      });
      let u = !1;
      const h = () => {
        u = !0;
      };
      this._map.once("click", h), this._map.once("dblclick", h), this._map.once("dragstart", h), this._map.once("mousedown", h), this._map.once("touchstart", h), this._map.once("wheel", h), this._map.once("moveend", () => {
        this._map.off("click", h), this._map.off("dblclick", h), this._map.off("dragstart", h), this._map.off("mousedown", h), this._map.off("touchstart", h), this._map.off("wheel", h), !u && (this.lastUpdatedCenter = this._map.getCenter());
      });
    });
    T(this, "_finishSetupUI", (t) => {
      if (this._map) {
        if (t === !1) {
          const o = this._map._getUIString("GeolocateControl.LocationNotAvailable");
          this._geolocateButton.disabled = !0, this._geolocateButton.title = o, this._geolocateButton.setAttribute("aria-label", o);
        } else {
          const o = this._map._getUIString("GeolocateControl.FindMyLocation");
          this._geolocateButton.disabled = !1, this._geolocateButton.title = o, this._geolocateButton.setAttribute("aria-label", o);
        }
        this.options.trackUserLocation && (this._geolocateButton.setAttribute("aria-pressed", "false"), this._watchState = "OFF"), this.options.showUserLocation && (this._dotElement = Z("div", "maplibregl-user-location-dot"), this._userLocationDotMarker = new Ae({ element: this._dotElement }), this._circleElement = Z("div", "maplibregl-user-location-accuracy-circle"), this._accuracyCircleMarker = new Ae({
          element: this._circleElement,
          pitchAlignment: "map"
        }), this.options.trackUserLocation && (this._watchState = "OFF"), this._map.on("move", this._onZoom)), this._geolocateButton.addEventListener("click", this.trigger.bind(this)), this._setup = !0, this.options.trackUserLocation && this._map.on("moveend", (o) => {
          const a = o.originalEvent && o.originalEvent.type === "resize", n = this.lastUpdatedCenter.distanceTo(this._map.getCenter());
          !o.geolocateSource && this._watchState === "ACTIVE_LOCK" && !a && n > 1 && (this._watchState = "BACKGROUND", this._geolocateButton.classList.add("maplibregl-ctrl-geolocate-background"), this._geolocateButton.classList.remove("maplibregl-ctrl-geolocate-active"), this.fire(new Event("trackuserlocationend")));
        });
      }
    });
    T(this, "_onZoom", () => {
      this.options.showUserLocation && this.options.showAccuracyCircle && this._updateCircleRadius();
    });
  }
  _updateCircleRadius() {
    if (this._watchState !== "BACKGROUND" && this._watchState !== "ACTIVE_LOCK")
      return;
    const t = [
      this._lastKnownPosition.coords.longitude,
      this._lastKnownPosition.coords.latitude
    ], o = this._map.project(t), a = this._map.unproject([o.x, o.y]), n = this._map.unproject([o.x + 20, o.y]), l = a.distanceTo(n) / 20, i = Math.ceil(2 * this._accuracy / l);
    this._circleElement.style.width = `${i}px`, this._circleElement.style.height = `${i}px`;
  }
}
var F, z, G, H, B, Y, x, Ue, N, Ge;
class le {
  constructor(e, t) {
    U(this, x);
    U(this, F);
    T(this, "map");
    U(this, z);
    U(this, G);
    U(this, H);
    U(this, B, !1);
    U(this, Y);
    e.style !== void 0 && $(this, B, !0), $(this, F, {
      // set defaults
      zoomAdjust: -4,
      position: "top-right",
      // inherit map options
      ...t,
      // override any lingering control options
      forceNoAttributionControl: !0,
      attributionControl: !1,
      navigationControl: !1,
      geolocateControl: !1,
      maptilerLogo: !1,
      minimap: !1,
      hash: !1,
      pitchAdjust: !1,
      // override map options with new user defined minimap options
      ...e,
      containerStyle: {
        border: "1px solid #000",
        width: "400px",
        height: "300px",
        ...e.containerStyle ?? {}
      }
    }), e.lockZoom !== void 0 && (L(this, F).minZoom = e.lockZoom, L(this, F).maxZoom = e.lockZoom);
  }
  setStyle(e, t) {
    L(this, B) || this.map.setStyle(e, t), k(this, x, N).call(this);
  }
  addLayer(e, t) {
    return L(this, B) || this.map.addLayer(e, t), k(this, x, N).call(this), this.map;
  }
  moveLayer(e, t) {
    return L(this, B) || this.map.moveLayer(e, t), k(this, x, N).call(this), this.map;
  }
  removeLayer(e) {
    return L(this, B) || this.map.removeLayer(e), k(this, x, N).call(this), this;
  }
  setLayerZoomRange(e, t, o) {
    return L(this, B) || this.map.setLayerZoomRange(e, t, o), k(this, x, N).call(this), this;
  }
  setFilter(e, t, o) {
    return L(this, B) || this.map.setFilter(e, t, o), k(this, x, N).call(this), this;
  }
  setPaintProperty(e, t, o, a) {
    return L(this, B) || this.map.setPaintProperty(e, t, o, a), k(this, x, N).call(this), this;
  }
  setLayoutProperty(e, t, o, a) {
    return L(this, B) || this.map.setLayoutProperty(e, t, o, a), k(this, x, N).call(this), this;
  }
  setGlyphs(e, t) {
    return L(this, B) || this.map.setGlyphs(e, t), k(this, x, N).call(this), this;
  }
  onAdd(e) {
    $(this, z, e), $(this, G, Z("div", "maplibregl-ctrl maplibregl-ctrl-group"));
    for (const [t, o] of Object.entries(L(this, F).containerStyle))
      L(this, G).style.setProperty(t, o);
    return L(this, F).container = L(this, G), L(this, F).zoom = e.getZoom() + L(this, F).zoomAdjust, this.map = new to(L(this, F)), this.map.once("style.load", () => {
      this.map.resize();
    }), this.map.once("load", () => {
      k(this, x, Ue).call(this, L(this, F).parentRect), $(this, Y, k(this, x, Ge).call(this));
    }), L(this, G);
  }
  onRemove() {
    var e;
    (e = L(this, Y)) == null || e.call(this), ze(L(this, G));
  }
}
F = new WeakMap(), z = new WeakMap(), G = new WeakMap(), H = new WeakMap(), B = new WeakMap(), Y = new WeakMap(), x = new WeakSet(), Ue = function(e) {
  e === void 0 || e.linePaint === void 0 && e.fillPaint === void 0 || ($(this, H, {
    type: "Feature",
    properties: {
      name: "parentRect"
    },
    geometry: {
      type: "Polygon",
      coordinates: [[[], [], [], [], []]]
    }
  }), this.map.addSource("parentRect", {
    type: "geojson",
    data: L(this, H)
  }), (e.lineLayout !== void 0 || e.linePaint !== void 0) && this.map.addLayer({
    id: "parentRectOutline",
    type: "line",
    source: "parentRect",
    layout: {
      ...e.lineLayout
    },
    paint: {
      "line-color": "#FFF",
      "line-width": 1,
      "line-opacity": 0.85,
      ...e.linePaint
    }
  }), e.fillPaint !== void 0 && this.map.addLayer({
    id: "parentRectFill",
    type: "fill",
    source: "parentRect",
    layout: {},
    paint: {
      "fill-color": "#08F",
      "fill-opacity": 0.135,
      ...e.fillPaint
    }
  }), k(this, x, N).call(this));
}, N = function() {
  if (L(this, H) === void 0) return;
  const { devicePixelRatio: e } = window, t = L(this, z).getCanvas(), o = t.width / e, a = t.height / e, n = L(this, z).unproject.bind(L(this, z)), l = n([0, 0]), i = n([o, 0]), u = n([0, a]), h = n([o, a]);
  L(this, H).geometry.coordinates = [
    [u.toArray(), h.toArray(), i.toArray(), l.toArray(), u.toArray()]
  ], this.map.getSource("parentRect").setData(L(this, H));
}, Ge = function() {
  const { pitchAdjust: e } = L(this, F), t = () => {
    l("parent");
  }, o = () => {
    l("minimap");
  }, a = () => {
    L(this, z).on("move", t), this.map.on("move", o);
  }, n = () => {
    L(this, z).off("move", t), this.map.off("move", o);
  }, l = (i) => {
    n();
    const u = i === "parent" ? L(this, z) : this.map, h = i === "parent" ? this.map : L(this, z), c = u.getCenter(), s = u.getZoom() + L(this, F).zoomAdjust * (i === "parent" ? 1 : -1), d = u.getBearing(), p = u.getPitch();
    h.jumpTo({
      center: c,
      zoom: s,
      bearing: d,
      pitch: e ? p : 0
    }), k(this, x, N).call(this), a();
  };
  return a(), () => {
    n();
  };
};
const eo = {
  POINT: "POINT",
  COUNTRY: "COUNTRY"
};
class to extends f.Map {
  constructor(t) {
    $t(t.container), t.apiKey && (M.apiKey = t.apiKey);
    const o = Se(t.style), a = location.hash;
    M.apiKey || console.warn("MapTiler Cloud API key is not set. Visit https://maptiler.com and try Cloud for free!");
    let n = {
      compact: !1
    };
    t.customAttribution ? n.customAttribution = t.customAttribution : t.attributionControl && typeof t.attributionControl == "object" && (n = {
      ...n,
      ...t.attributionControl
    });
    super({
      ...t,
      style: o,
      maplibreLogo: !1,
      transformRequest: Ce(t.transformRequest),
      attributionControl: t.forceNoAttributionControl === !0 ? !1 : n
    });
    T(this, "isTerrainEnabled", !1);
    T(this, "terrainExaggeration", 1);
    T(this, "primaryLanguage");
    T(this, "terrainGrowing", !1);
    T(this, "terrainFlattening", !1);
    T(this, "minimap");
    T(this, "forceLanguageUpdate");
    T(this, "languageAlwaysBeenStyle");
    T(this, "isReady", !1);
    M.caching && !ce && console.warn(
      "The cache API is only available in secure contexts. More info at https://developer.mozilla.org/en-US/docs/Web/API/Cache"
    ), M.caching && ce && Dt(), this.primaryLanguage = t.language ?? M.primaryLanguage, this.forceLanguageUpdate = !(this.primaryLanguage === C.STYLE || this.primaryLanguage === C.STYLE_LOCK), this.languageAlwaysBeenStyle = this.primaryLanguage === C.STYLE, this.terrainExaggeration = t.terrainExaggeration ?? this.terrainExaggeration, this.once("styledata", async () => {
      if (!t.geolocate || t.center || t.hash && a)
        return;
      try {
        if (t.geolocate === eo.COUNTRY) {
          await this.fitToIpBounds();
          return;
        }
      } catch (d) {
        console.warn(d.message);
      }
      let c;
      try {
        await this.centerOnIpPoint(t.zoom), c = this.getCameraHash();
      } catch (d) {
        console.warn(d.message);
      }
      (await navigator.permissions.query({
        name: "geolocation"
      })).state === "granted" && navigator.geolocation.getCurrentPosition(
        // success callback
        (d) => {
          c === this.getCameraHash() && (this.terrain ? this.easeTo({
            center: [d.coords.longitude, d.coords.latitude],
            zoom: t.zoom || 12,
            duration: 2e3
          }) : this.once("terrain", () => {
            this.easeTo({
              center: [d.coords.longitude, d.coords.latitude],
              zoom: t.zoom || 12,
              duration: 2e3
            });
          }));
        },
        // error callback
        null,
        // options
        {
          maximumAge: 24 * 3600 * 1e3,
          // a day in millisec
          timeout: 5e3,
          // milliseconds
          enableHighAccuracy: !1
        }
      );
    }), this.on("styledata", () => {
      this.setPrimaryLanguage(this.primaryLanguage);
    }), this.on("styledata", () => {
      this.getTerrain() === null && this.isTerrainEnabled && this.enableTerrain(this.terrainExaggeration);
    }), this.once("load", async () => {
      let c = { logo: null };
      try {
        const s = Object.keys(this.style.sourceCaches).map((m) => this.getSource(m)).filter(
          (m) => m && "url" in m && typeof m.url == "string" && (m == null ? void 0 : m.url.includes("tiles.json"))
        ), d = new URL(s[0].url);
        d.searchParams.has("key") || d.searchParams.append("key", M.apiKey), c = await (await fetch(d.href)).json();
      } catch {
      }
      if (t.forceNoAttributionControl !== !0)
        if ("logo" in c && c.logo) {
          const s = c.logo;
          this.addControl(new be({ logoURL: s }), t.logoPosition);
        } else t.maptilerLogo && this.addControl(new be(), t.logoPosition);
      if (t.scaleControl) {
        const s = t.scaleControl === !0 || t.scaleControl === void 0 ? "bottom-right" : t.scaleControl, d = new Wt({ unit: M.unit });
        this.addControl(d, s), M.on("unit", (p) => {
          d.setUnit(p);
        });
      }
      if (t.navigationControl !== !1) {
        const s = t.navigationControl === !0 || t.navigationControl === void 0 ? "top-right" : t.navigationControl;
        this.addControl(new Jt(), s);
      }
      if (t.geolocateControl !== !1) {
        const s = t.geolocateControl === !0 || t.geolocateControl === void 0 ? "top-right" : t.geolocateControl;
        this.addControl(
          // new maplibregl.GeolocateControl({
          new Qt({
            positionOptions: {
              enableHighAccuracy: !0,
              maximumAge: 0,
              timeout: 6e3
            },
            fitBoundsOptions: {
              maxZoom: 15
            },
            trackUserLocation: !0,
            showAccuracyCircle: !0,
            showUserLocation: !0
          }),
          s
        );
      }
      if (t.terrainControl) {
        const s = t.terrainControl === !0 || t.terrainControl === void 0 ? "top-right" : t.terrainControl;
        this.addControl(new qt(), s);
      }
      if (t.fullscreenControl) {
        const s = t.fullscreenControl === !0 || t.fullscreenControl === void 0 ? "top-right" : t.fullscreenControl;
        this.addControl(new Yt({}), s);
      }
      this.isReady = !0, this.fire("ready", { target: this });
    });
    let l = !1, i = !1, u;
    this.once("ready", () => {
      l = !0, i && this.fire("loadWithTerrain", u);
    }), this.once("style.load", () => {
      const { minimap: c } = t;
      if (typeof c == "object") {
        const {
          zoom: s,
          center: d,
          style: p,
          language: m,
          apiKey: y,
          maptilerLogo: b,
          antialias: P,
          refreshExpiredTiles: w,
          maxBounds: _,
          scrollZoom: E,
          minZoom: R,
          maxZoom: D,
          boxZoom: V,
          locale: W,
          fadeDuration: q,
          crossSourceCollisions: J,
          clickTolerance: Xe,
          bounds: Qe,
          fitBoundsOptions: et,
          pixelRatio: tt,
          validateStyle: ot
        } = t;
        this.minimap = new le(c, {
          zoom: s,
          center: d,
          style: p,
          language: m,
          apiKey: y,
          container: "null",
          maptilerLogo: b,
          antialias: P,
          refreshExpiredTiles: w,
          maxBounds: _,
          scrollZoom: E,
          minZoom: R,
          maxZoom: D,
          boxZoom: V,
          locale: W,
          fadeDuration: q,
          crossSourceCollisions: J,
          clickTolerance: Xe,
          bounds: Qe,
          fitBoundsOptions: et,
          pixelRatio: tt,
          validateStyle: ot
        }), this.addControl(this.minimap, c.position ?? "bottom-left");
      } else c === !0 ? (this.minimap = new le({}, t), this.addControl(this.minimap, "bottom-left")) : c !== void 0 && c !== !1 && (this.minimap = new le({}, t), this.addControl(this.minimap, c));
    });
    const h = (c) => {
      c.terrain && (i = !0, u = {
        type: "loadWithTerrain",
        target: this,
        terrain: c.terrain
      }, this.off("terrain", h), l && this.fire("loadWithTerrain", u));
    };
    this.on("terrain", h), t.terrain && this.enableTerrain(t.terrainExaggeration ?? this.terrainExaggeration);
  }
  /**
   * Awaits for _this_ Map instance to be "loaded" and returns a Promise to the Map.
   * If _this_ Map instance is already loaded, the Promise is resolved directly,
   * otherwise, it is resolved as a result of the "load" event.
   * @returns
   */
  async onLoadAsync() {
    return new Promise((t) => {
      if (this.loaded())
        return t(this);
      this.once("load", () => {
        t(this);
      });
    });
  }
  /**
   * Awaits for _this_ Map instance to be "ready" and returns a Promise to the Map.
   * If _this_ Map instance is already ready, the Promise is resolved directly,
   * otherwise, it is resolved as a result of the "ready" event.
   * A map instance is "ready" when all the controls that can be managed by the contructor are
   * dealt with. This happens after the "load" event, due to the asynchronous nature
   * of some built-in controls.
   */
  async onReadyAsync() {
    return new Promise((t) => {
      if (this.isReady)
        return t(this);
      this.once("ready", () => {
        t(this);
      });
    });
  }
  /**
   * Awaits for _this_ Map instance to be "loaded" as well as with terrain being non-null for the first time
   * and returns a Promise to the Map.
   * If _this_ Map instance is already loaded with terrain, the Promise is resolved directly,
   * otherwise, it is resolved as a result of the "loadWithTerrain" event.
   * @returns
   */
  async onLoadWithTerrainAsync() {
    return new Promise((t) => {
      if (this.isReady && this.terrain)
        return t(this);
      this.once("loadWithTerrain", () => {
        t(this);
      });
    });
  }
  /**
   * Update the style of the map.
   * Can be:
   * - a full style URL (possibly with API key)
   * - a shorthand with only the MapTIler style name (eg. `"streets-v2"`)
   * - a longer form with the prefix `"maptiler://"` (eg. `"maptiler://streets-v2"`)
   */
  setStyle(t, o) {
    var a;
    return (a = this.minimap) == null || a.setStyle(t), this.forceLanguageUpdate = !0, this.once("idle", () => {
      this.forceLanguageUpdate = !1;
    }), super.setStyle(Se(t), o);
  }
  /**
   * Adds a [MapLibre style layer](https://maplibre.org/maplibre-style-spec/layers)
   * to the map's style.
   *
   * A layer defines how data from a specified source will be styled. Read more about layer types
   * and available paint and layout properties in the [MapLibre Style Specification](https://maplibre.org/maplibre-style-spec/layers).
   *
   * @param layer - The layer to add,
   * conforming to either the MapLibre Style Specification's [layer definition](https://maplibre.org/maplibre-style-spec/layers) or,
   * less commonly, the {@link CustomLayerInterface} specification.
   * The MapLibre Style Specification's layer definition is appropriate for most layers.
   *
   * @param beforeId - The ID of an existing layer to insert the new layer before,
   * resulting in the new layer appearing visually beneath the existing layer.
   * If this argument is not specified, the layer will be appended to the end of the layers array
   * and appear visually above all other layers.
   *
   * @returns `this`
   */
  addLayer(t, o) {
    var a;
    return (a = this.minimap) == null || a.addLayer(t, o), super.addLayer(t, o);
  }
  /**
   * Moves a layer to a different z-position.
   *
   * @param id - The ID of the layer to move.
   * @param beforeId - The ID of an existing layer to insert the new layer before. When viewing the map, the `id` layer will appear beneath the `beforeId` layer. If `beforeId` is omitted, the layer will be appended to the end of the layers array and appear above all other layers on the map.
   * @returns `this`
   *
   * @example
   * Move a layer with ID 'polygon' before the layer with ID 'country-label'. The `polygon` layer will appear beneath the `country-label` layer on the map.
   * ```ts
   * map.moveLayer('polygon', 'country-label');
   * ```
   */
  moveLayer(t, o) {
    var a;
    return (a = this.minimap) == null || a.moveLayer(t, o), super.moveLayer(t, o);
  }
  /**
   * Removes the layer with the given ID from the map's style.
   *
   * An {@link ErrorEvent} will be fired if the image parameter is invald.
   *
   * @param id - The ID of the layer to remove
   * @returns `this`
   *
   * @example
   * If a layer with ID 'state-data' exists, remove it.
   * ```ts
   * if (map.getLayer('state-data')) map.removeLayer('state-data');
   * ```
   */
  removeLayer(t) {
    var o;
    return (o = this.minimap) == null || o.removeLayer(t), super.removeLayer(t);
  }
  /**
   * Sets the zoom extent for the specified style layer. The zoom extent includes the
   * [minimum zoom level](https://maplibre.org/maplibre-style-spec/layers/#minzoom)
   * and [maximum zoom level](https://maplibre.org/maplibre-style-spec/layers/#maxzoom))
   * at which the layer will be rendered.
   *
   * Note: For style layers using vector sources, style layers cannot be rendered at zoom levels lower than the
   * minimum zoom level of the _source layer_ because the data does not exist at those zoom levels. If the minimum
   * zoom level of the source layer is higher than the minimum zoom level defined in the style layer, the style
   * layer will not be rendered at all zoom levels in the zoom range.
   */
  setLayerZoomRange(t, o, a) {
    var n;
    return (n = this.minimap) == null || n.setLayerZoomRange(t, o, a), super.setLayerZoomRange(t, o, a);
  }
  /**
   * Sets the filter for the specified style layer.
   *
   * Filters control which features a style layer renders from its source.
   * Any feature for which the filter expression evaluates to `true` will be
   * rendered on the map. Those that are false will be hidden.
   *
   * Use `setFilter` to show a subset of your source data.
   *
   * To clear the filter, pass `null` or `undefined` as the second parameter.
   */
  setFilter(t, o, a) {
    var n;
    return (n = this.minimap) == null || n.setFilter(t, o, a), super.setFilter(t, o, a);
  }
  /**
   * Sets the value of a paint property in the specified style layer.
   *
   * @param layerId - The ID of the layer to set the paint property in.
   * @param name - The name of the paint property to set.
   * @param value - The value of the paint property to set.
   * Must be of a type appropriate for the property, as defined in the [MapLibre Style Specification](https://maplibre.org/maplibre-style-spec/).
   * @param options - Options object.
   * @returns `this`
   * @example
   * ```ts
   * map.setPaintProperty('my-layer', 'fill-color', '#faafee');
   * ```
   */
  setPaintProperty(t, o, a, n) {
    var l;
    return (l = this.minimap) == null || l.setPaintProperty(t, o, a, n), super.setPaintProperty(t, o, a, n);
  }
  /**
   * Sets the value of a layout property in the specified style layer.
   * Layout properties define how the layer is styled.
   * Layout properties for layers of the same type are documented together.
   * Layers of different types have different layout properties.
   * See the [MapLibre Style Specification](https://maplibre.org/maplibre-style-spec/) for the complete list of layout properties.
   * @param layerId - The ID of the layer to set the layout property in.
   * @param name - The name of the layout property to set.
   * @param value - The value of the layout property to set.
   * Must be of a type appropriate for the property, as defined in the [MapLibre Style Specification](https://maplibre.org/maplibre-style-spec/).
   * @param options - Options object.
   * @returns `this`
   */
  setLayoutProperty(t, o, a, n) {
    var l;
    return (l = this.minimap) == null || l.setLayoutProperty(t, o, a, n), super.setLayoutProperty(t, o, a, n);
  }
  /**
   * Sets the value of the style's glyphs property.
   *
   * @param glyphsUrl - Glyph URL to set. Must conform to the [MapLibre Style Specification](https://maplibre.org/maplibre-style-spec/glyphs/).
   * @param options - Options object.
   * @returns `this`
   * @example
   * ```ts
   * map.setGlyphs('https://demotiles.maplibre.org/font/{fontstack}/{range}.pbf');
   * ```
   */
  setGlyphs(t, o) {
    var a;
    return (a = this.minimap) == null || a.setGlyphs(t, o), super.setGlyphs(t, o);
  }
  getStyleLanguage() {
    return !this.style.stylesheet.metadata || typeof this.style.stylesheet.metadata != "object" ? null : "maptiler:language" in this.style.stylesheet.metadata && typeof this.style.stylesheet.metadata["maptiler:language"] == "string" ? dt(this.style.stylesheet.metadata["maptiler:language"]) : null;
  }
  /**
   * Define the primary language of the map. Note that not all the languages shorthands provided are available.
   */
  setLanguage(t) {
    var o, a;
    (a = (o = this.minimap) == null ? void 0 : o.map) == null || a.setLanguage(t), this.onStyleReady(() => {
      this.setPrimaryLanguage(t);
    });
  }
  /**
   * Define the primary language of the map. Note that not all the languages shorthands provided are available.
   */
  setPrimaryLanguage(t) {
    const o = this.getStyleLanguage(), a = pt(t, C);
    if (!a) {
      console.warn(`The language "${a}" is not supported.`);
      return;
    }
    if (!(a.flag === C.STYLE.flag && o && (o.flag === C.AUTO.flag || o.flag === C.VISITOR.flag)) && (a.flag !== C.STYLE.flag && (this.languageAlwaysBeenStyle = !1), this.languageAlwaysBeenStyle || this.primaryLanguage === a && !this.forceLanguageUpdate))
      return;
    if (this.primaryLanguage.flag === C.STYLE_LOCK.flag) {
      console.warn(
        "The language cannot be changed because this map has been instantiated with the STYLE_LOCK language flag."
      );
      return;
    }
    this.primaryLanguage = a;
    let n = a;
    if (a.flag === C.STYLE.flag) {
      if (!o) {
        console.warn("The style has no default languages or has an invalid one.");
        return;
      }
      n = o;
    }
    let l = C.LOCAL.flag, i = `{${l}}`;
    n.flag === C.VISITOR.flag ? (l = ye().flag, i = [
      "case",
      ["all", ["has", l], ["has", C.LOCAL.flag]],
      [
        "case",
        ["==", ["get", l], ["get", C.LOCAL.flag]],
        ["get", C.LOCAL.flag],
        [
          "format",
          ["get", l],
          { "font-scale": 0.8 },
          `
`,
          ["get", C.LOCAL.flag],
          { "font-scale": 1.1 }
        ]
      ],
      ["get", C.LOCAL.flag]
    ]) : n.flag === C.VISITOR_ENGLISH.flag ? (l = C.ENGLISH.flag, i = [
      "case",
      ["all", ["has", l], ["has", C.LOCAL.flag]],
      [
        "case",
        ["==", ["get", l], ["get", C.LOCAL.flag]],
        ["get", C.LOCAL.flag],
        [
          "format",
          ["get", l],
          { "font-scale": 0.8 },
          `
`,
          ["get", C.LOCAL.flag],
          { "font-scale": 1.1 }
        ]
      ],
      ["get", C.LOCAL.flag]
    ]) : n.flag === C.AUTO.flag ? (l = ye().flag, i = ["case", ["has", l], ["get", l], ["get", C.LOCAL.flag]]) : n === C.LOCAL ? (l = C.LOCAL.flag, i = `{${l}}`) : (l = n.flag, i = ["case", ["has", l], ["get", l], ["get", C.LOCAL.flag]]);
    const { layers: u } = this.getStyle();
    for (const h of u) {
      if (h.type !== "symbol")
        continue;
      const c = h, s = this.getSource(c.source);
      if (!s || !("url" in s && typeof s.url == "string") || new URL(s.url).host !== O.maptilerApiHost)
        continue;
      const { id: p, layout: m } = c;
      if (!m || !("text-field" in m))
        continue;
      const y = this.getLayoutProperty(p, "text-field");
      typeof y == "string" && (y.toLowerCase().includes("ref") || y.toLowerCase().includes("housenumber")) || this.setLayoutProperty(p, "text-field", i);
    }
  }
  /**
   * Get the primary language
   * @returns
   */
  getPrimaryLanguage() {
    return this.primaryLanguage;
  }
  /**
   * Get the exaggeration factor applied to the terrain
   * @returns
   */
  getTerrainExaggeration() {
    return this.terrainExaggeration;
  }
  /**
   * Know if terrian is enabled or not
   * @returns
   */
  hasTerrain() {
    return this.isTerrainEnabled;
  }
  growTerrain(t, o = 1e3) {
    if (!this.terrain)
      return;
    const a = performance.now(), n = this.terrain.exaggeration, l = t - n, i = () => {
      if (!this.terrain || this.terrainFlattening)
        return;
      const u = (performance.now() - a) / o;
      if (u < 0.99) {
        const h = 1 - (1 - u) ** 4, c = n + h * l;
        this.terrain.exaggeration = c, requestAnimationFrame(i);
      } else
        this.terrainGrowing = !1, this.terrainFlattening = !1, this.terrain.exaggeration = t;
      this._elevationFreeze = !1, this.triggerRepaint();
    };
    this.terrainGrowing = !0, this.terrainFlattening = !1, requestAnimationFrame(i);
  }
  /**
   * Enables the 3D terrain visualization
   */
  enableTerrain(t = this.terrainExaggeration) {
    if (t < 0) {
      console.warn("Terrain exaggeration cannot be negative.");
      return;
    }
    const o = async (n) => {
      !this.terrain || n.type !== "data" || n.dataType !== "source" || !("source" in n) || n.sourceId !== "maptiler-terrain" || n.source.type !== "raster-dem" || n.isSourceLoaded && (this.off("data", o), this.growTerrain(t));
    }, a = () => {
      this.isTerrainEnabled = !0, this.terrainExaggeration = t, this.on("data", o), this.addSource(O.terrainSourceId, {
        type: "raster-dem",
        url: O.terrainSourceURL
      }), this.setTerrain({
        source: O.terrainSourceId,
        exaggeration: 0
      });
    };
    if (this.getTerrain()) {
      this.isTerrainEnabled = !0, this.growTerrain(t);
      return;
    }
    this.loaded() || this.isTerrainEnabled ? a() : this.once("load", () => {
      this.getTerrain() && this.getSource(O.terrainSourceId) || a();
    });
  }
  /**
   * Disable the 3D terrain visualization
   */
  disableTerrain() {
    if (!this.terrain)
      return;
    this.isTerrainEnabled = !1;
    const t = 1 * 1e3, o = performance.now(), a = this.terrain.exaggeration, n = () => {
      if (!this.terrain || this.terrainGrowing)
        return;
      const l = (performance.now() - o) / t;
      if (this._elevationFreeze = !1, l < 0.99) {
        const i = (1 - l) ** 4, u = a * i;
        this.terrain.exaggeration = u, requestAnimationFrame(n);
      } else
        this.terrain.exaggeration = 0, this.terrainGrowing = !1, this.terrainFlattening = !1, this.setTerrain(), this.getSource(O.terrainSourceId) && this.removeSource(O.terrainSourceId);
      this.triggerRepaint();
    };
    this.terrainGrowing = !1, this.terrainFlattening = !0, requestAnimationFrame(n);
  }
  /**
   * Sets the 3D terrain exageration factor.
   * If the terrain was not enabled prior to the call of this method,
   * the method `.enableTerrain()` will be called.
   * If `animate` is `true`, the terrain transformation will be animated in the span of 1 second.
   * If `animate` is `false`, no animated transition to the newly defined exaggeration.
   */
  setTerrainExaggeration(t, o = !0) {
    !o && this.terrain ? (this.terrainExaggeration = t, this.terrain.exaggeration = t, this.triggerRepaint()) : this.enableTerrain(t);
  }
  /**
   * Perform an action when the style is ready. It could be at the moment of calling this method
   * or later.
   */
  onStyleReady(t) {
    this.isStyleLoaded() ? t() : this.once("styledata", () => {
      t();
    });
  }
  async fitToIpBounds() {
    const t = await ge.info();
    this.fitBounds(t.country_bounds, {
      duration: 0,
      padding: 100
    });
  }
  async centerOnIpPoint(t) {
    const o = await ge.info();
    this.jumpTo({
      center: [(o == null ? void 0 : o.longitude) ?? 0, (o == null ? void 0 : o.latitude) ?? 0],
      zoom: t || 11
    });
  }
  getCameraHash() {
    const t = new Float32Array(5), o = this.getCenter();
    return t[0] = o.lng, t[1] = o.lat, t[2] = this.getZoom(), t[3] = this.getPitch(), t[4] = this.getBearing(), ft.fromUint8Array(new Uint8Array(t.buffer));
  }
  /**
   * Get the SDK config object.
   * This is convenient to dispatch the SDK configuration to externally built layers
   * that do not directly have access to the SDK configuration but do have access to a Map instance.
   */
  getSdkConfig() {
    return M;
  }
  /**
   * Get the MapTiler session ID. Convenient to dispatch to externaly built component
   * that do not directly have access to the SDK configuration but do have access to a Map instance.
   * @returns
   */
  getMaptilerSessionId() {
    return Be;
  }
  /**
   *  Updates the requestManager's transform request with a new function.
   *
   * @param transformRequest A callback run before the Map makes a request for an external URL. The callback can be used to modify the url, set headers, or set the credentials property for cross-origin requests.
   *    Expected to return an object with a `url` property and optionally `headers` and `credentials` properties
   *
   * @returns {Map} `this`
   *
   *  @example
   *  map.setTransformRequest((url: string, resourceType: string) => {});
   */
  setTransformRequest(t) {
    return super.setTransformRequest(Ce(t)), this;
  }
}
function de(r) {
  if (typeof DOMParser < "u") {
    const e = new DOMParser().parseFromString(r, "application/xml");
    if (e.querySelector("parsererror"))
      throw new Error("The provided string is not valid XML");
    return e;
  }
  throw new Error("No XML parser found");
}
function He(r, e) {
  if (!r.hasChildNodes())
    return !1;
  for (const t of Array.from(r.childNodes)) {
    const o = t.nodeName;
    if (typeof o == "string" && o.trim().toLowerCase() === e.toLowerCase())
      return !0;
  }
  return !1;
}
function Re(r) {
  if (typeof XMLSerializer < "u")
    return new XMLSerializer().serializeToString(r);
  throw new Error("No XML serializer found");
}
function Ve(r) {
  const e = typeof r == "string" ? de(r) : r;
  if (!He(e, "gpx"))
    throw new Error("The XML document is not valid GPX");
  const t = I(e, "trk"), o = I(e, "rte"), a = I(e, "wpt"), n = {
    type: "FeatureCollection",
    features: []
  };
  for (const l of Array.from(t)) {
    const i = no(l);
    i && n.features.push(i);
  }
  for (const l of Array.from(o)) {
    const i = lo(l);
    i && n.features.push(i);
  }
  for (const l of Array.from(a))
    n.features.push(io(l));
  return n;
}
function $e(r, e) {
  let t = r;
  if (typeof t == "string" && (t = de(t)), !He(t, "kml"))
    throw new Error("The XML document is not valid KML");
  const o = {
    type: "FeatureCollection",
    features: []
  }, a = {}, n = {}, l = {}, i = I(t, "Placemark"), u = I(t, "Style"), h = I(t, "StyleMap");
  for (const c of Array.from(u)) {
    const s = Ee(e !== void 0 ? e(c) : Re(c)).toString(16);
    a[`#${j(c, "id")}`] = s, n[s] = c;
  }
  for (const c of Array.from(h)) {
    a[`#${j(c, "id")}`] = Ee(
      e !== void 0 ? e(c) : Re(c)
    ).toString(16);
    const s = I(c, "Pair"), d = {};
    for (const p of Array.from(s))
      d[S(v(p, "key")) ?? ""] = S(v(p, "styleUrl"));
    l[`#${j(c, "id")}`] = d;
  }
  for (const c of Array.from(i))
    o.features = o.features.concat(ao(c, a, n, l));
  return o;
}
function _e(r) {
  if (r === null) return ["#000000", 1];
  let e = "", t = 1, o = r;
  return o.substring(0, 1) === "#" && (o = o.substring(1)), (o.length === 6 || o.length === 3) && (e = o), o.length === 8 && (t = Number.parseInt(o.substring(0, 2), 16) / 255, e = `#${o.substring(6, 8)}${o.substring(4, 6)}${o.substring(2, 4)}`), [e ?? "#000000", t ?? 1];
}
function oo(r) {
  return Ze(r.split(" "));
}
function ro(r) {
  let e = I(r, "coord");
  const t = [], o = [];
  e.length === 0 && (e = I(r, "gx:coord"));
  for (const n of Array.from(e))
    t.push(oo(S(n) ?? ""));
  const a = I(r, "when");
  for (const n of Array.from(a)) o.push(S(n));
  return {
    coords: t,
    times: o
  };
}
function Q(r) {
  const e = ["Polygon", "LineString", "Point", "Track", "gx:Track"];
  let t, o, a, n, l;
  const i = [], u = [];
  if (v(r, "MultiGeometry") !== null)
    return Q(v(r, "MultiGeometry"));
  if (v(r, "MultiTrack") !== null)
    return Q(v(r, "MultiTrack"));
  if (v(r, "gx:MultiTrack") !== null)
    return Q(v(r, "gx:MultiTrack"));
  for (a = 0; a < e.length; a++)
    if (o = I(r, e[a]), o) {
      for (n = 0; n < o.length; n++)
        if (t = o[n], e[a] === "Point")
          i.push({
            type: "Point",
            coordinates: We(S(v(t, "coordinates")) ?? "")
          });
        else if (e[a] === "LineString")
          i.push({
            type: "LineString",
            coordinates: Ie(S(v(t, "coordinates")) ?? "")
          });
        else if (e[a] === "Polygon") {
          const h = I(t, "LinearRing"), c = [];
          for (l = 0; l < h.length; l++)
            c.push(Ie(S(v(h[l], "coordinates")) ?? ""));
          i.push({
            type: "Polygon",
            coordinates: c
          });
        } else if (e[a] === "Track" || e[a] === "gx:Track") {
          const h = ro(t);
          i.push({
            type: "LineString",
            coordinates: h.coords
          }), h.times.length && u.push(h.times);
        }
    }
  return { geoms: i, coordTimes: u };
}
function ao(r, e, t, o) {
  const a = Q(r), n = {}, l = S(v(r, "name")), i = S(v(r, "address")), u = S(v(r, "description")), h = v(r, "TimeSpan"), c = v(r, "TimeStamp"), s = v(r, "ExtendedData"), d = v(r, "visibility");
  let p, m = S(v(r, "styleUrl")), y = v(r, "LineStyle"), b = v(r, "PolyStyle");
  if (!a.geoms.length) return [];
  if (l && (n.name = l), i && (n.address = i), m) {
    m[0] !== "#" && (m = `#${m}`), n.styleUrl = m, e[m] && (n.styleHash = e[m]), o[m] && (n.styleMapHash = o[m], n.styleHash = e[o[m].normal ?? ""]);
    const w = t[n.styleHash ?? ""];
    if (w) {
      y || (y = v(w, "LineStyle")), b || (b = v(w, "PolyStyle"));
      const _ = v(w, "IconStyle");
      if (_) {
        const E = v(_, "Icon");
        if (E) {
          const R = S(v(E, "href"));
          R && (n.icon = R);
        }
      }
    }
  }
  if (u && (n.description = u), h) {
    const w = S(v(h, "begin")), _ = S(v(h, "end"));
    w && _ && (n.timespan = { begin: w, end: _ });
  }
  if (c !== null && (n.timestamp = S(v(c, "when")) ?? (/* @__PURE__ */ new Date()).toISOString()), y !== null) {
    const w = _e(S(v(y, "color"))), _ = w[0], E = w[1], R = Number.parseFloat(S(v(y, "width")) ?? "");
    _ && (n.stroke = _), Number.isNaN(E) || (n["stroke-opacity"] = E), Number.isNaN(R) || (n["stroke-width"] = R);
  }
  if (b) {
    const w = _e(S(v(b, "color"))), _ = w[0], E = w[1], R = S(v(b, "fill")), D = S(v(b, "outline"));
    _ && (n.fill = _), Number.isNaN(E) || (n["fill-opacity"] = E), R && (n["fill-opacity"] = R === "1" ? n["fill-opacity"] || 1 : 0), D && (n["stroke-opacity"] = D === "1" ? n["stroke-opacity"] || 1 : 0);
  }
  if (s) {
    const w = I(s, "Data"), _ = I(s, "SimpleData");
    for (p = 0; p < w.length; p++)
      n[w[p].getAttribute("name") ?? ""] = S(v(w[p], "value")) ?? "";
    for (p = 0; p < _.length; p++)
      n[_[p].getAttribute("name") ?? ""] = S(_[p]) ?? "";
  }
  d !== null && (n.visibility = S(d) ?? ""), a.coordTimes.length !== 0 && (n.coordTimes = a.coordTimes.length === 1 ? a.coordTimes[0] : a.coordTimes);
  const P = {
    type: "Feature",
    geometry: a.geoms.length === 1 ? a.geoms[0] : {
      type: "GeometryCollection",
      geometries: a.geoms
    },
    properties: n
  };
  return j(r, "id") && (P.id = j(r, "id") ?? void 0), [P];
}
function je(r, e) {
  const t = I(r, e), o = [], a = [];
  let n = [];
  const l = t.length;
  if (!(l < 2)) {
    for (let i = 0; i < l; i++) {
      const u = Ye(t[i]);
      o.push(u.coordinates), u.time && a.push(u.time), (u.heartRate || n.length) && (n.length === 0 && (n = new Array(i).fill(null)), n.push(u.heartRate));
    }
    return {
      line: o,
      times: a,
      heartRates: n
    };
  }
}
function no(r) {
  const e = I(r, "trkseg"), t = [], o = [], a = [];
  let n;
  for (let i = 0; i < e.length; i++)
    if (n = je(e[i], "trkpt"), n !== void 0 && (n.line && t.push(n.line), n.times && n.times.length && o.push(n.times), a.length || n.heartRates && n.heartRates.length)) {
      if (!a.length)
        for (let u = 0; u < i; u++)
          a.push(new Array(t[u].length).fill(null));
      n.heartRates && n.heartRates.length ? a.push(n.heartRates) : a.push(new Array(n.line.length).fill(null));
    }
  if (t.length === 0) return;
  const l = {
    ...pe(r),
    ...Ke(v(r, "extensions"))
  };
  return o.length !== 0 && (l.coordTimes = t.length === 1 ? o[0] : o), a.length !== 0 && (l.heartRates = t.length === 1 ? a[0] : a), t.length === 1 ? {
    type: "Feature",
    properties: l,
    geometry: {
      type: "LineString",
      coordinates: t[0]
    }
  } : {
    type: "Feature",
    properties: l,
    geometry: {
      type: "MultiLineString",
      coordinates: t
    }
  };
}
function lo(r) {
  const e = je(r, "rtept");
  return e === void 0 ? void 0 : {
    type: "Feature",
    properties: {
      ...pe(r),
      ...Ke(v(r, "extensions"))
    },
    geometry: {
      type: "LineString",
      coordinates: e.line
    }
  };
}
function io(r) {
  return {
    type: "Feature",
    properties: { ...pe(r), ...ue(r, ["sym"]) },
    geometry: {
      type: "Point",
      coordinates: Ye(r).coordinates
    }
  };
}
function Ke(r) {
  const e = {};
  if (r) {
    const t = v(r, "line");
    if (t) {
      const o = S(v(t, "color")), a = Number.parseFloat(S(v(t, "opacity")) ?? "0"), n = Number.parseFloat(S(v(t, "width")) ?? "0");
      o && (e.stroke = o), Number.isNaN(a) || (e["stroke-opacity"] = a), Number.isNaN(n) || (e["stroke-width"] = n * 96 / 25.4);
    }
  }
  return e;
}
function pe(r) {
  const e = ue(r, [
    "name",
    "cmt",
    "desc",
    "type",
    "time",
    "keywords"
  ]), t = I(r, "link");
  if (t.length !== 0) {
    e.links = [];
    for (const o of Array.from(t)) {
      const a = {
        href: j(o, "href"),
        ...ue(o, ["text", "type"])
      };
      e.links.push(a);
    }
  }
  return e;
}
function Ee(r) {
  let e = 0;
  if (!r || !r.length) return e;
  for (let t = 0; t < r.length; t++)
    e = (e << 5) - e + r.charCodeAt(t) | 0;
  return e;
}
function I(r, e) {
  return r.getElementsByTagName(e);
}
function j(r, e) {
  return r.getAttribute(e);
}
function xe(r, e) {
  return Number.parseFloat(j(r, e) ?? "0");
}
function v(r, e) {
  const t = I(r, e);
  return t.length ? t[0] : null;
}
function so(r) {
  return r.normalize && r.normalize(), r;
}
function Ze(r) {
  return r.map(Number.parseFloat).map((e) => Number.isNaN(e) ? null : e);
}
function S(r) {
  return r && so(r), r && r.textContent;
}
function ue(r, e) {
  const t = {};
  let o, a;
  for (a = 0; a < e.length; a++)
    o = v(r, e[a]), o && (t[e[a]] = S(o) ?? "");
  return t;
}
function We(r) {
  return Ze(r.replace(/\s*/g, "").split(","));
}
function Ie(r) {
  const e = r.replace(/^\s*|\s*$/g, "").split(/\s+/), t = [];
  for (const o of e) t.push(We(o));
  return t;
}
function Ye(r) {
  const e = [xe(r, "lon"), xe(r, "lat")], t = v(r, "ele"), o = v(r, "gpxtpx:hr") || v(r, "hr"), a = v(r, "time");
  let n;
  return t && (n = Number.parseFloat(S(t) ?? "0"), Number.isNaN(n) || e.push(n)), {
    coordinates: e,
    time: a ? S(a) : null,
    heartRate: o !== null ? Number.parseFloat(S(o) ?? "0") : null
  };
}
function co(r) {
  let e = r;
  try {
    typeof e == "string" && (e = de(e));
  } catch {
    return null;
  }
  try {
    return Ve(e);
  } catch {
  }
  try {
    return $e(e);
  } catch {
  }
  return null;
}
function X(r) {
  const e = r.toString(16);
  return e.length === 1 ? `0${e}` : e;
}
function uo(r) {
  return `#${X(r[0])}${X(r[1])}${X(r[2])}${r.length === 4 ? X(r[3]) : ""}`;
}
class g extends Array {
  constructor(t = {}) {
    super();
    T(this, "min", 0);
    T(this, "max", 1);
    "min" in t && (this.min = t.min), "max" in t && (this.max = t.max), "stops" in t && this.setStops(t.stops, { clone: !1 });
  }
  /**
   * Converts a array-definition color ramp definition into a usable ColorRamp instance.
   * Note: units are not converted and may need to to be converted beforehand (eg. kelvin to centigrade)
   * @param cr
   * @returns
   */
  static fromArrayDefinition(t) {
    return new g({
      stops: t.map((o) => ({
        value: o[0],
        color: o[1]
      }))
    });
  }
  setStops(t, o = { clone: !0 }) {
    const a = o.clone ? this.clone() : this;
    a.length = 0;
    let n = Number.POSITIVE_INFINITY, l = Number.NEGATIVE_INFINITY;
    for (let i = 0; i < t.length; i += 1)
      n = Math.min(n, t[i].value), l = Math.max(l, t[i].value), a.push({
        value: t[i].value,
        color: t[i].color.slice()
        // we want to make sure we do a deep copy and not a reference
      });
    return a.sort((i, u) => i.value < u.value ? -1 : 1), this.min = n, this.max = l, a;
  }
  scale(t, o, a = { clone: !0 }) {
    const n = a.clone, l = this[0].value, u = this.at(-1).value - l, h = o - t, c = [];
    for (let s = 0; s < this.length; s += 1) {
      const m = (this[s].value - l) / u * h + t;
      n ? c.push({
        value: m,
        color: this[s].color.slice()
      }) : this[s].value = m;
    }
    return n ? new g({ stops: c }) : this;
  }
  // for some reason, I had to reimplement this
  at(t) {
    return t < 0 ? this[this.length + t] : this[t];
  }
  clone() {
    return new g({ stops: this.getRawColorStops() });
  }
  getRawColorStops() {
    const t = [];
    for (let o = 0; o < this.length; o += 1)
      t.push({ value: this[o].value, color: this[o].color });
    return t;
  }
  reverse(t = { clone: !0 }) {
    const o = t.clone ? this.clone() : this;
    for (let a = 0; a < ~~(o.length / 2); a += 1) {
      const n = o[a].color;
      o[a].color = o.at(-(a + 1)).color, o.at(-(a + 1)).color = n;
    }
    return o;
  }
  getBounds() {
    return { min: this.min, max: this.max };
  }
  getColor(t, o = { smooth: !0 }) {
    if (t <= this[0].value)
      return this[0].color;
    if (t >= this.at(-1).value)
      return this.at(-1).color;
    for (let a = 0; a < this.length - 1; a += 1) {
      if (t > this[a + 1].value)
        continue;
      const n = this[a].color;
      if (!o.smooth)
        return n.slice();
      const l = this[a].value, i = this[a + 1].value, u = this[a + 1].color, h = (i - t) / (i - l);
      return n.map(
        (c, s) => Math.round(c * h + u[s] * (1 - h))
      );
    }
    return [0, 0, 0];
  }
  /**
   * Get the color as an hexadecimal string
   */
  getColorHex(t, o = {
    smooth: !0,
    withAlpha: !1
  }) {
    return uo(this.getColor(t, o));
  }
  /**
   * Get the color of the color ramp at a relative position in [0, 1]
   */
  getColorRelative(t, o = { smooth: !0 }) {
    const a = this.getBounds();
    return this.getColor(a.min + t * (a.max - a.min), o);
  }
  getCanvasStrip(t = {
    horizontal: !0,
    size: 512,
    smooth: !0
  }) {
    const o = document.createElement("canvas");
    o.width = t.horizontal ? t.size : 1, o.height = t.horizontal ? 1 : t.size;
    const a = o.getContext("2d");
    if (!a) throw new Error("Canvs context is missing");
    const n = a.getImageData(0, 0, o.width, o.height), l = n.data, i = t.size, u = this[0].value, s = (this.at(-1).value - u) / i;
    for (let d = 0; d < i; d += 1) {
      const p = this.getColor(u + d * s, {
        smooth: t.smooth
      });
      l[d * 4] = p[0], l[d * 4 + 1] = p[1], l[d * 4 + 2] = p[2], l[d * 4 + 3] = p.length > 3 ? p[3] : 255;
    }
    return a.putImageData(n, 0, 0), o;
  }
  /**
   * Apply a non-linear ressampling. This will create a new instance of ColorRamp with the same bounds.
   */
  resample(t, o = 15) {
    const a = this.getBounds(), n = this.scale(0, 1), l = 1 / (o - 1);
    let i;
    if (t === "ease-in-square")
      i = Array.from({ length: o }, (c, s) => {
        const d = s * l, p = d ** 2, m = n.getColor(p);
        return { value: d, color: m };
      });
    else if (t === "ease-out-square")
      i = Array.from({ length: o }, (c, s) => {
        const d = s * l, p = 1 - (1 - d) ** 2, m = n.getColor(p);
        return { value: d, color: m };
      });
    else if (t === "ease-out-sqrt")
      i = Array.from({ length: o }, (c, s) => {
        const d = s * l, p = d ** 0.5, m = n.getColor(p);
        return { value: d, color: m };
      });
    else if (t === "ease-in-sqrt")
      i = Array.from({ length: o }, (c, s) => {
        const d = s * l, p = 1 - (1 - d) ** 0.5, m = n.getColor(p);
        return { value: d, color: m };
      });
    else if (t === "ease-out-exp")
      i = Array.from({ length: o }, (c, s) => {
        const d = s * l, p = 1 - 2 ** (-10 * d), m = n.getColor(p);
        return { value: d, color: m };
      });
    else if (t === "ease-in-exp")
      i = Array.from({ length: o }, (c, s) => {
        const d = s * l, p = 2 ** (10 * d - 10), m = n.getColor(p);
        return { value: d, color: m };
      });
    else
      throw new Error("Invalid ressampling method.");
    return new g({ stops: i }).scale(a.min, a.max);
  }
  /**
   * Makes a clone of this color ramp that is fully transparant at the begining of their range
   */
  transparentStart() {
    const t = this.getRawColorStops();
    t.unshift({
      value: t[0].value,
      color: t[0].color.slice()
    }), t[1].value += 1e-3;
    for (const o of t)
      o.color.length === 3 && o.color.push(255);
    return t[0].color[3] = 0, new g({ stops: t });
  }
  /**
   * Check if this color ramp has a transparent start
   */
  hasTransparentStart() {
    return this[0].color.length === 4 && this[0].color[3] === 0;
  }
}
const qe = {
  /**
   * A fully transparent [0, 0, 0, 0] colorramp to hide data.
   * Defined in interval [0, 1], without unit.
   */
  NULL: new g({
    stops: [
      { value: 0, color: [0, 0, 0, 0] },
      { value: 1, color: [0, 0, 0, 0] }
    ]
  }),
  GRAY: new g({
    stops: [
      { value: 0, color: [0, 0, 0] },
      { value: 1, color: [255, 255, 255] }
    ]
  }),
  /**
   * Classic jet color ramp.
   * Defined in interval [0, 1], without unit.
   */
  JET: new g({
    stops: [
      { value: 0, color: [0, 0, 131] },
      { value: 0.125, color: [0, 60, 170] },
      { value: 0.375, color: [5, 255, 255] },
      { value: 0.625, color: [255, 255, 0] },
      { value: 0.875, color: [250, 0, 0] },
      { value: 1, color: [128, 0, 0] }
    ]
  }),
  /**
   * Classic HSV color ramp (hue, saturation, value).
   * Defined in interval [0, 1], without unit.
   */
  HSV: new g({
    stops: [
      { value: 0, color: [255, 0, 0] },
      { value: 0.169, color: [253, 255, 2] },
      { value: 0.173, color: [247, 255, 2] },
      { value: 0.337, color: [0, 252, 4] },
      { value: 0.341, color: [0, 252, 10] },
      { value: 0.506, color: [1, 249, 255] },
      { value: 0.671, color: [2, 0, 253] },
      { value: 0.675, color: [8, 0, 253] },
      { value: 0.839, color: [255, 0, 251] },
      { value: 0.843, color: [255, 0, 245] },
      { value: 1, color: [255, 0, 6] }
    ]
  }),
  /**
   * Classic hot color ramp.
   * Defined in interval [0, 1], without unit.
   */
  HOT: new g({
    stops: [
      { value: 0, color: [0, 0, 0] },
      { value: 0.3, color: [230, 0, 0] },
      { value: 0.6, color: [255, 210, 0] },
      { value: 1, color: [255, 255, 255] }
    ]
  }),
  /**
   * Classic spring color ramp.
   * Defined in interval [0, 1], without unit.
   */
  SPRING: new g({
    stops: [
      { value: 0, color: [255, 0, 255] },
      { value: 1, color: [255, 255, 0] }
    ]
  }),
  /**
   * Classic summer color ramp.
   * Defined in interval [0, 1], without unit.
   */
  SUMMER: new g({
    stops: [
      { value: 0, color: [0, 128, 102] },
      { value: 1, color: [255, 255, 102] }
    ]
  }),
  /**
   * Classic autommn color ramp.
   * Defined in interval [0, 1], without unit.
   */
  AUTOMN: new g({
    stops: [
      { value: 0, color: [255, 0, 0] },
      { value: 1, color: [255, 255, 0] }
    ]
  }),
  /**
   * Classic winter color ramp.
   * Defined in interval [0, 1], without unit.
   */
  WINTER: new g({
    stops: [
      { value: 0, color: [0, 0, 255] },
      { value: 1, color: [0, 255, 128] }
    ]
  }),
  /**
   * Classic bone color ramp.
   * Defined in interval [0, 1], without unit.
   */
  BONE: new g({
    stops: [
      { value: 0, color: [0, 0, 0] },
      { value: 0.376, color: [84, 84, 116] },
      { value: 0.753, color: [169, 200, 200] },
      { value: 1, color: [255, 255, 255] }
    ]
  }),
  /**
   * Classic copper color ramp.
   * Defined in interval [0, 1], without unit.
   */
  COPPER: new g({
    stops: [
      { value: 0, color: [0, 0, 0] },
      { value: 0.804, color: [255, 160, 102] },
      { value: 1, color: [255, 199, 127] }
    ]
  }),
  /**
   * Classic greys color ramp.
   * Defined in interval [0, 1], without unit.
   */
  GREYS: new g({
    stops: [
      { value: 0, color: [0, 0, 0] },
      { value: 1, color: [255, 255, 255] }
    ]
  }),
  /**
   * Classic yignbu color ramp (blue to light yellow).
   * Defined in interval [0, 1], without unit.
   */
  YIGNBU: new g({
    stops: [
      { value: 0, color: [8, 29, 88] },
      { value: 0.125, color: [37, 52, 148] },
      { value: 0.25, color: [34, 94, 168] },
      { value: 0.375, color: [29, 145, 192] },
      { value: 0.5, color: [65, 182, 196] },
      { value: 0.625, color: [127, 205, 187] },
      { value: 0.75, color: [199, 233, 180] },
      { value: 0.875, color: [237, 248, 217] },
      { value: 1, color: [255, 255, 217] }
    ]
  }),
  /**
   * Classic greens color ramp.
   * Defined in interval [0, 1], without unit.
   */
  GREENS: new g({
    stops: [
      { value: 0, color: [0, 68, 27] },
      { value: 0.125, color: [0, 109, 44] },
      { value: 0.25, color: [35, 139, 69] },
      { value: 0.375, color: [65, 171, 93] },
      { value: 0.5, color: [116, 196, 118] },
      { value: 0.625, color: [161, 217, 155] },
      { value: 0.75, color: [199, 233, 192] },
      { value: 0.875, color: [229, 245, 224] },
      { value: 1, color: [247, 252, 245] }
    ]
  }),
  /**
   * Classic yiorrd color ramp (red to light yellow).
   * Defined in interval [0, 1], without unit.
   */
  YIORRD: new g({
    stops: [
      { value: 0, color: [128, 0, 38] },
      { value: 0.125, color: [189, 0, 38] },
      { value: 0.25, color: [227, 26, 28] },
      { value: 0.375, color: [252, 78, 42] },
      { value: 0.5, color: [253, 141, 60] },
      { value: 0.625, color: [254, 178, 76] },
      { value: 0.75, color: [254, 217, 118] },
      { value: 0.875, color: [255, 237, 160] },
      { value: 1, color: [255, 255, 204] }
    ]
  }),
  /**
   * Classic blue-red color ramp.
   * Defined in interval [0, 1], without unit.
   */
  BLUERED: new g({
    stops: [
      { value: 0, color: [0, 0, 255] },
      { value: 1, color: [255, 0, 0] }
    ]
  }),
  /**
   * Classic rdbu color ramp.
   * Defined in interval [0, 1], without unit.
   */
  RDBU: new g({
    stops: [
      { value: 0, color: [5, 10, 172] },
      { value: 0.35, color: [106, 137, 247] },
      { value: 0.5, color: [190, 190, 190] },
      { value: 0.6, color: [220, 170, 132] },
      { value: 0.7, color: [230, 145, 90] },
      { value: 1, color: [178, 10, 28] }
    ]
  }),
  /**
   * Classic picnic color ramp.
   * Defined in interval [0, 1], without unit.
   */
  PICNIC: new g({
    stops: [
      { value: 0, color: [0, 0, 255] },
      { value: 0.1, color: [51, 153, 255] },
      { value: 0.2, color: [102, 204, 255] },
      { value: 0.3, color: [153, 204, 255] },
      { value: 0.4, color: [204, 204, 255] },
      { value: 0.5, color: [255, 255, 255] },
      { value: 0.6, color: [255, 204, 255] },
      { value: 0.7, color: [255, 153, 255] },
      { value: 0.8, color: [255, 102, 204] },
      { value: 0.9, color: [255, 102, 102] },
      { value: 1, color: [255, 0, 0] }
    ]
  }),
  /**
   * Classic rainbow color ramp.
   * Defined in interval [0, 1], without unit.
   */
  RAINBOW: new g({
    stops: [
      { value: 0, color: [150, 0, 90] },
      { value: 0.125, color: [0, 0, 200] },
      { value: 0.25, color: [0, 25, 255] },
      { value: 0.375, color: [0, 152, 255] },
      { value: 0.5, color: [44, 255, 150] },
      { value: 0.625, color: [151, 255, 0] },
      { value: 0.75, color: [255, 234, 0] },
      { value: 0.875, color: [255, 111, 0] },
      { value: 1, color: [255, 0, 0] }
    ]
  }),
  /**
   * Classic Portland color ramp.
   * Defined in interval [0, 1], without unit.
   */
  PORTLAND: new g({
    stops: [
      { value: 0, color: [12, 51, 131] },
      { value: 0.25, color: [10, 136, 186] },
      { value: 0.5, color: [242, 211, 56] },
      { value: 0.75, color: [242, 143, 56] },
      { value: 1, color: [217, 30, 30] }
    ]
  }),
  /**
   * Classic blackbody color ramp.
   * Defined in interval [0, 1], without unit.
   */
  BLACKBODY: new g({
    stops: [
      { value: 0, color: [0, 0, 0] },
      { value: 0.2, color: [230, 0, 0] },
      { value: 0.4, color: [230, 210, 0] },
      { value: 0.7, color: [255, 255, 255] },
      { value: 1, color: [160, 200, 255] }
    ]
  }),
  /**
   * Classic earth color ramp.
   * Defined in interval [0, 1], without unit.
   */
  EARTH: new g({
    stops: [
      { value: 0, color: [0, 0, 130] },
      { value: 0.1, color: [0, 180, 180] },
      { value: 0.2, color: [40, 210, 40] },
      { value: 0.4, color: [230, 230, 50] },
      { value: 0.6, color: [120, 70, 20] },
      { value: 1, color: [255, 255, 255] }
    ]
  }),
  /**
   * Classic electric color ramp.
   * Defined in interval [0, 1], without unit.
   */
  ELECTRIC: new g({
    stops: [
      { value: 0, color: [0, 0, 0] },
      { value: 0.15, color: [30, 0, 100] },
      { value: 0.4, color: [120, 0, 100] },
      { value: 0.6, color: [160, 90, 0] },
      { value: 0.8, color: [230, 200, 0] },
      { value: 1, color: [255, 250, 220] }
    ]
  }),
  /**
   * Classic viridis color ramp.
   * Defined in interval [0, 1], without unit.
   */
  VIRIDIS: new g({
    stops: [
      { value: 0, color: [68, 1, 84] },
      { value: 0.13, color: [71, 44, 122] },
      { value: 0.25, color: [59, 81, 139] },
      { value: 0.38, color: [44, 113, 142] },
      { value: 0.5, color: [33, 144, 141] },
      { value: 0.63, color: [39, 173, 129] },
      { value: 0.75, color: [92, 200, 99] },
      { value: 0.88, color: [170, 220, 50] },
      { value: 1, color: [253, 231, 37] }
    ]
  }),
  /**
   * Classic inferno color ramp.
   * Defined in interval [0, 1], without unit.
   */
  INFERNO: new g({
    stops: [
      { value: 0, color: [0, 0, 4] },
      { value: 0.13, color: [31, 12, 72] },
      { value: 0.25, color: [85, 15, 109] },
      { value: 0.38, color: [136, 34, 106] },
      { value: 0.5, color: [186, 54, 85] },
      { value: 0.63, color: [227, 89, 51] },
      { value: 0.75, color: [249, 140, 10] },
      { value: 0.88, color: [249, 201, 50] },
      { value: 1, color: [252, 255, 164] }
    ]
  }),
  /**
   * Classic magma color ramp.
   * Defined in interval [0, 1], without unit.
   */
  MAGMA: new g({
    stops: [
      { value: 0, color: [0, 0, 4] },
      { value: 0.13, color: [28, 16, 68] },
      { value: 0.25, color: [79, 18, 123] },
      { value: 0.38, color: [129, 37, 129] },
      { value: 0.5, color: [181, 54, 122] },
      { value: 0.63, color: [229, 80, 100] },
      { value: 0.75, color: [251, 135, 97] },
      { value: 0.88, color: [254, 194, 135] },
      { value: 1, color: [252, 253, 191] }
    ]
  }),
  /**
   * Classic plasma color ramp.
   * Defined in interval [0, 1], without unit.
   */
  PLASMA: new g({
    stops: [
      { value: 0, color: [13, 8, 135] },
      { value: 0.13, color: [75, 3, 161] },
      { value: 0.25, color: [125, 3, 168] },
      { value: 0.38, color: [168, 34, 150] },
      { value: 0.5, color: [203, 70, 121] },
      { value: 0.63, color: [229, 107, 93] },
      { value: 0.75, color: [248, 148, 65] },
      { value: 0.88, color: [253, 195, 40] },
      { value: 1, color: [240, 249, 33] }
    ]
  }),
  /**
   * Classic warm color ramp.
   * Defined in interval [0, 1], without unit.
   */
  WARM: new g({
    stops: [
      { value: 0, color: [125, 0, 179] },
      { value: 0.13, color: [172, 0, 187] },
      { value: 0.25, color: [219, 0, 170] },
      { value: 0.38, color: [255, 0, 130] },
      { value: 0.5, color: [255, 63, 74] },
      { value: 0.63, color: [255, 123, 0] },
      { value: 0.75, color: [234, 176, 0] },
      { value: 0.88, color: [190, 228, 0] },
      { value: 1, color: [147, 255, 0] }
    ]
  }),
  /**
   * Classic cool color ramp.
   * Defined in interval [0, 1], without unit.
   */
  COOL: new g({
    stops: [
      { value: 0, color: [125, 0, 179] },
      { value: 0.13, color: [116, 0, 218] },
      { value: 0.25, color: [98, 74, 237] },
      { value: 0.38, color: [68, 146, 231] },
      { value: 0.5, color: [0, 204, 197] },
      { value: 0.63, color: [0, 247, 146] },
      { value: 0.75, color: [0, 255, 88] },
      { value: 0.88, color: [40, 255, 8] },
      { value: 1, color: [147, 255, 0] }
    ]
  }),
  /**
   * Classic rainboz soft color ramp.
   * Defined in interval [0, 1], without unit.
   */
  RAINBOW_SOFT: new g({
    stops: [
      { value: 0, color: [125, 0, 179] },
      { value: 0.1, color: [199, 0, 180] },
      { value: 0.2, color: [255, 0, 121] },
      { value: 0.3, color: [255, 108, 0] },
      { value: 0.4, color: [222, 194, 0] },
      { value: 0.5, color: [150, 255, 0] },
      { value: 0.6, color: [0, 255, 55] },
      { value: 0.7, color: [0, 246, 150] },
      { value: 0.8, color: [50, 167, 222] },
      { value: 0.9, color: [103, 51, 235] },
      { value: 1, color: [124, 0, 186] }
    ]
  }),
  /**
   * Classic bathymetry color ramp.
   * Defined in interval [0, 1], without unit.
   */
  BATHYMETRY: new g({
    stops: [
      { value: 0, color: [40, 26, 44] },
      { value: 0.13, color: [59, 49, 90] },
      { value: 0.25, color: [64, 76, 139] },
      { value: 0.38, color: [63, 110, 151] },
      { value: 0.5, color: [72, 142, 158] },
      { value: 0.63, color: [85, 174, 163] },
      { value: 0.75, color: [120, 206, 163] },
      { value: 0.88, color: [187, 230, 172] },
      { value: 1, color: [253, 254, 204] }
    ]
  }),
  /**
   * Classic cdom color ramp.
   * Defined in interval [0, 1], without unit.
   */
  CDOM: new g({
    stops: [
      { value: 0, color: [47, 15, 62] },
      { value: 0.13, color: [87, 23, 86] },
      { value: 0.25, color: [130, 28, 99] },
      { value: 0.38, color: [171, 41, 96] },
      { value: 0.5, color: [206, 67, 86] },
      { value: 0.63, color: [230, 106, 84] },
      { value: 0.75, color: [242, 149, 103] },
      { value: 0.88, color: [249, 193, 135] },
      { value: 1, color: [254, 237, 176] }
    ]
  }),
  /**
   * Classic chlorophyll color ramp.
   * Defined in interval [0, 1], without unit.
   */
  CHLOROPHYLL: new g({
    stops: [
      { value: 0, color: [18, 36, 20] },
      { value: 0.13, color: [25, 63, 41] },
      { value: 0.25, color: [24, 91, 59] },
      { value: 0.38, color: [13, 119, 72] },
      { value: 0.5, color: [18, 148, 80] },
      { value: 0.63, color: [80, 173, 89] },
      { value: 0.75, color: [132, 196, 122] },
      { value: 0.88, color: [175, 221, 162] },
      { value: 1, color: [215, 249, 208] }
    ]
  }),
  /**
   * Classic density color ramp.
   * Defined in interval [0, 1], without unit.
   */
  DENSITY: new g({
    stops: [
      { value: 0, color: [54, 14, 36] },
      { value: 0.13, color: [89, 23, 80] },
      { value: 0.25, color: [110, 45, 132] },
      { value: 0.38, color: [120, 77, 178] },
      { value: 0.5, color: [120, 113, 213] },
      { value: 0.63, color: [115, 151, 228] },
      { value: 0.75, color: [134, 185, 227] },
      { value: 0.88, color: [177, 214, 227] },
      { value: 1, color: [230, 241, 241] }
    ]
  }),
  /**
   * Classic freesurface blue color ramp.
   * Defined in interval [0, 1], without unit.
   */
  FREESURFACE_BLUE: new g({
    stops: [
      { value: 0, color: [30, 4, 110] },
      { value: 0.13, color: [47, 14, 176] },
      { value: 0.25, color: [41, 45, 236] },
      { value: 0.38, color: [25, 99, 212] },
      { value: 0.5, color: [68, 131, 200] },
      { value: 0.63, color: [114, 156, 197] },
      { value: 0.75, color: [157, 181, 203] },
      { value: 0.88, color: [200, 208, 216] },
      { value: 1, color: [241, 237, 236] }
    ]
  }),
  /**
   * Classic freesurface red color ramp.
   * Defined in interval [0, 1], without unit.
   */
  FREESURFACE_RED: new g({
    stops: [
      { value: 0, color: [60, 9, 18] },
      { value: 0.13, color: [100, 17, 27] },
      { value: 0.25, color: [142, 20, 29] },
      { value: 0.38, color: [177, 43, 27] },
      { value: 0.5, color: [192, 87, 63] },
      { value: 0.63, color: [205, 125, 105] },
      { value: 0.75, color: [216, 162, 148] },
      { value: 0.88, color: [227, 199, 193] },
      { value: 1, color: [241, 237, 236] }
    ]
  }),
  /**
   * Classic oxygen color ramp.
   * Defined in interval [0, 1], without unit.
   */
  OXYGEN: new g({
    stops: [
      { value: 0, color: [64, 5, 5] },
      { value: 0.13, color: [106, 6, 15] },
      { value: 0.25, color: [144, 26, 7] },
      { value: 0.38, color: [168, 64, 3] },
      { value: 0.5, color: [188, 100, 4] },
      { value: 0.63, color: [206, 136, 11] },
      { value: 0.75, color: [220, 174, 25] },
      { value: 0.88, color: [231, 215, 44] },
      { value: 1, color: [248, 254, 105] }
    ]
  }),
  /**
   * Classic par color ramp.
   * Defined in interval [0, 1], without unit.
   */
  PAR: new g({
    stops: [
      { value: 0, color: [51, 20, 24] },
      { value: 0.13, color: [90, 32, 35] },
      { value: 0.25, color: [129, 44, 34] },
      { value: 0.38, color: [159, 68, 25] },
      { value: 0.5, color: [182, 99, 19] },
      { value: 0.63, color: [199, 134, 22] },
      { value: 0.75, color: [212, 171, 35] },
      { value: 0.88, color: [221, 210, 54] },
      { value: 1, color: [225, 253, 75] }
    ]
  }),
  /**
   * Classic phase color ramp.
   * Defined in interval [0, 1], without unit.
   */
  PHASE: new g({
    stops: [
      { value: 0, color: [145, 105, 18] },
      { value: 0.13, color: [184, 71, 38] },
      { value: 0.25, color: [186, 58, 115] },
      { value: 0.38, color: [160, 71, 185] },
      { value: 0.5, color: [110, 97, 218] },
      { value: 0.63, color: [50, 123, 164] },
      { value: 0.75, color: [31, 131, 110] },
      { value: 0.88, color: [77, 129, 34] },
      { value: 1, color: [145, 105, 18] }
    ]
  }),
  /**
   * Classic salinity color ramp.
   * Defined in interval [0, 1], without unit.
   */
  SALINITY: new g({
    stops: [
      { value: 0, color: [42, 24, 108] },
      { value: 0.13, color: [33, 50, 162] },
      { value: 0.25, color: [15, 90, 145] },
      { value: 0.38, color: [40, 118, 137] },
      { value: 0.5, color: [59, 146, 135] },
      { value: 0.63, color: [79, 175, 126] },
      { value: 0.75, color: [120, 203, 104] },
      { value: 0.88, color: [193, 221, 100] },
      { value: 1, color: [253, 239, 154] }
    ]
  }),
  /**
   * Classic temperature color ramp.
   * Defined in interval [0, 1], without unit.
   */
  TEMPERATURE: new g({
    stops: [
      { value: 0, color: [4, 35, 51] },
      { value: 0.13, color: [23, 51, 122] },
      { value: 0.25, color: [85, 59, 157] },
      { value: 0.38, color: [129, 79, 143] },
      { value: 0.5, color: [175, 95, 130] },
      { value: 0.63, color: [222, 112, 101] },
      { value: 0.75, color: [249, 146, 66] },
      { value: 0.88, color: [249, 196, 65] },
      { value: 1, color: [232, 250, 91] }
    ]
  }),
  /**
   * Classic turbidity color ramp.
   * Defined in interval [0, 1], without unit.
   */
  TURBIDITY: new g({
    stops: [
      { value: 0, color: [34, 31, 27] },
      { value: 0.13, color: [65, 50, 41] },
      { value: 0.25, color: [98, 69, 52] },
      { value: 0.38, color: [131, 89, 57] },
      { value: 0.5, color: [161, 112, 59] },
      { value: 0.63, color: [185, 140, 66] },
      { value: 0.75, color: [202, 174, 88] },
      { value: 0.88, color: [216, 209, 126] },
      { value: 1, color: [233, 246, 171] }
    ]
  }),
  /**
   * Classic velocity blue color ramp.
   * Defined in interval [0, 1], without unit.
   */
  VELOCITY_BLUE: new g({
    stops: [
      { value: 0, color: [17, 32, 64] },
      { value: 0.13, color: [35, 52, 116] },
      { value: 0.25, color: [29, 81, 156] },
      { value: 0.38, color: [31, 113, 162] },
      { value: 0.5, color: [50, 144, 169] },
      { value: 0.63, color: [87, 173, 176] },
      { value: 0.75, color: [149, 196, 189] },
      { value: 0.88, color: [203, 221, 211] },
      { value: 1, color: [254, 251, 230] }
    ]
  }),
  /**
   * Classic velocity green color ramp.
   * Defined in interval [0, 1], without unit.
   */
  VELOCITY_GREEN: new g({
    stops: [
      { value: 0, color: [23, 35, 19] },
      { value: 0.13, color: [24, 64, 38] },
      { value: 0.25, color: [11, 95, 45] },
      { value: 0.38, color: [39, 123, 35] },
      { value: 0.5, color: [95, 146, 12] },
      { value: 0.63, color: [152, 165, 18] },
      { value: 0.75, color: [201, 186, 69] },
      { value: 0.88, color: [233, 216, 137] },
      { value: 1, color: [255, 253, 205] }
    ]
  }),
  /**
   * Classic cube helix color ramp.
   * Defined in interval [0, 1], without unit.
   */
  CUBEHELIX: new g({
    stops: [
      { value: 0, color: [0, 0, 0] },
      { value: 0.07, color: [22, 5, 59] },
      { value: 0.13, color: [60, 4, 105] },
      { value: 0.2, color: [109, 1, 135] },
      { value: 0.27, color: [161, 0, 147] },
      { value: 0.33, color: [210, 2, 142] },
      { value: 0.4, color: [251, 11, 123] },
      { value: 0.47, color: [255, 29, 97] },
      { value: 0.53, color: [255, 54, 69] },
      { value: 0.6, color: [255, 85, 46] },
      { value: 0.67, color: [255, 120, 34] },
      { value: 0.73, color: [255, 157, 37] },
      { value: 0.8, color: [241, 191, 57] },
      { value: 0.87, color: [224, 220, 93] },
      { value: 0.93, color: [218, 241, 142] },
      { value: 1, color: [227, 253, 198] }
    ]
  }),
  /**
   * The cividis color ramp is color blind friendly.
   * Read more here https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0199239
   * Defined in interval [0, 1], without unit.
   */
  CIVIDIS: new g({
    stops: [
      { value: 0, color: [0, 32, 77, 255] },
      { value: 0.125, color: [5, 54, 110, 255] },
      { value: 0.25, color: [65, 77, 108, 255] },
      { value: 0.375, color: [97, 100, 111, 255] },
      { value: 0.5, color: [125, 124, 121, 255] },
      { value: 0.625, color: [156, 149, 120, 255] },
      { value: 0.75, color: [190, 175, 111, 255] },
      { value: 0.875, color: [225, 204, 94, 255] },
      { value: 1, color: [255, 235, 70, 255] }
    ]
  }),
  /**
   * Classic turbo color ramp.
   * This is a luminance-constant alternative to the jet, making it more
   * clor-blind friendly.
   * Defined in interval [0, 1], without unit.
   */
  TURBO: new g({
    stops: [
      { value: 0, color: [48, 18, 59, 255] },
      { value: 0.125, color: [70, 107, 227, 255] },
      { value: 0.25, color: [40, 187, 236, 255] },
      { value: 0.375, color: [49, 242, 153, 255] },
      { value: 0.5, color: [162, 252, 60, 255] },
      { value: 0.625, color: [237, 208, 58, 255] },
      { value: 0.75, color: [251, 128, 34, 255] },
      { value: 0.875, color: [210, 49, 5, 255] },
      { value: 1, color: [122, 4, 3, 255] }
    ]
  }),
  /**
   * The rocket color ramp is perceptually uniform, which makes it more
   * color bliend friendly than the classic magma color ramp.
   * Defined in interval [0, 1], without unit.
   */
  ROCKET: new g({
    stops: [
      { value: 0, color: [250, 235, 221, 0] },
      { value: 0.133, color: [250, 235, 221, 255] },
      { value: 0.266, color: [246, 170, 130, 255] },
      { value: 0.4, color: [240, 96, 67, 255] },
      { value: 0.533, color: [203, 27, 79, 255] },
      { value: 0.666, color: [132, 30, 90, 255] },
      { value: 0.8, color: [63, 27, 68, 255] },
      { value: 1, color: [3, 5, 26, 255] }
    ]
  }),
  /**
   * The mako color ramp is perceptually uniform and can be seen as
   * a color blind friendly alternative to bathymetry or yignbu.
   * Defined in interval [0, 1], without unit.
   */
  MAKO: new g({
    stops: [
      { value: 0, color: [11, 4, 5, 255] },
      { value: 0.125, color: [43, 28, 53, 255] },
      { value: 0.25, color: [62, 53, 107, 255] },
      { value: 0.375, color: [59, 86, 152, 255] },
      { value: 0.5, color: [53, 123, 162, 255] },
      { value: 0.625, color: [53, 158, 170, 255] },
      { value: 0.75, color: [73, 193, 173, 255] },
      { value: 0.875, color: [150, 221, 181, 255] },
      { value: 1, color: [222, 245, 229, 255] }
    ]
  })
};
async function po(r, e = {}) {
  const t = e.download ?? !1, o = await ho(r);
  if (t) {
    const a = e.filename ?? "maptiler_screenshot.png", n = document.createElement("a");
    n.style.display = "none", document.body.appendChild(n), n.href = URL.createObjectURL(o), n.download = a, n.click(), setTimeout(() => {
      document.body.removeChild(n), URL.revokeObjectURL(n.href);
    }, 0);
  }
  return o;
}
function ho(r) {
  return new Promise((e, t) => {
    r.redraw(), r.once("idle", () => {
      r.getCanvas().toBlob((o) => {
        if (!o)
          return t(Error("Screenshot could not be created."));
        e(o);
      }, "image/png");
    });
  });
}
const Me = [
  // https://colorhunt.co/palette/1d5b79468b97ef6262f3aa60
  ["#1D5B79", "#468B97", "#EF6262", "#F3AA60"],
  // https://colorhunt.co/palette/614bc333bbc585e6c5c8ffe0
  ["#614BC3", "#33BBC5", "#85E6C5", "#C8FFE0"],
  // https://colorhunt.co/palette/4619597a316fcd6688aed8cc
  ["#461959", "#7A316F", "#CD6688", "#AED8CC"],
  // https://colorhunt.co/palette/0079ff00dfa2f6fa70ff0060
  ["#0079FF", "#00DFA2", "#F6FA70", "#FF0060"],
  //https://colorhunt.co/palette/39b5e0a31acbff78f0f5ea5a
  ["#39B5E0", "#A31ACB", "#FF78F0", "#F5EA5A"],
  // https://colorhunt.co/palette/37e2d5590696c70a80fbcb0a
  ["#37E2D5", "#590696", "#C70A80", "#FBCB0A"],
  // https://colorhunt.co/palette/ffd36efff56d99ffcd9fb4ff
  ["#FFD36E", "#FFF56D", "#99FFCD", "#9FB4FF"],
  // https://colorhunt.co/palette/00ead3fff5b7ff449f005f99
  ["#00EAD3", "#FFF5B7", "#FF449F", "#005F99"],
  // https://colorhunt.co/palette/10a19d540375ff7000ffbf00
  ["#10A19D", "#540375", "#FF7000", "#FFBF00"]
];
function he() {
  return Me[~~(Math.random() * Me.length)][~~(Math.random() * 4)];
}
function te() {
  return `maptiler_source_${De()}`;
}
function oe() {
  return `maptiler_layer_${De()}`;
}
function Pe(r, e) {
  if (e <= r[0].zoom)
    return r[0].value;
  if (e >= r[r.length - 1].zoom)
    return r[r.length - 1].value;
  for (let t = 0; t < r.length - 1; t += 1)
    if (e >= r[t].zoom && e < r[t + 1].zoom) {
      const o = r[t + 1].zoom - r[t].zoom, a = (e - r[t].zoom) / o;
      return a * r[t + 1].value + (1 - a) * r[t].value;
    }
  return 0;
}
function K(r) {
  return ["interpolate", ["linear"], ["zoom"], ...r.flatMap((e) => [e.zoom, e.value])];
}
function A(r) {
  return ["interpolate", ["linear"], ["zoom"], ...r.flatMap((e) => [e.zoom, e.value])];
}
function mo(r, e) {
  if (typeof e == "number" && typeof r == "number")
    return 2 * e + r;
  if (typeof e == "number" && Array.isArray(r))
    return ["interpolate", ["linear"], ["zoom"], ...r.flatMap((t) => [t.zoom, 2 * e + t.value])];
  if (typeof r == "number" && Array.isArray(e))
    return ["interpolate", ["linear"], ["zoom"], ...e.flatMap((t) => [t.zoom, 2 * t.value + r])];
  if (Array.isArray(r) && Array.isArray(e)) {
    const t = Array.from(
      /* @__PURE__ */ new Set([...r.map((o) => o.zoom), ...e.map((o) => o.zoom)])
    ).sort((o, a) => o < a ? -1 : 1);
    return [
      "interpolate",
      ["linear"],
      ["zoom"],
      ...t.flatMap((o) => [o, 2 * Pe(e, o) + Pe(r, o)])
    ];
  }
  return 0;
}
function fo(r, e) {
  return ["interpolate", ["linear"], ["get", e], ...r.flatMap((t) => [t.propertyValue, t.value])];
}
function Je(r) {
  const e = r.trimStart(), t = `${e}${" ".repeat(r.length - e.length)}`, o = Array.from(t);
  if (!o.every((i) => i === " " || i === "_"))
    throw new Error("A dash pattern must be composed only of whitespace and underscore characters.");
  if (!(o.some((i) => i === "_") && o.some((i) => i === " ")))
    throw new Error("A dash pattern must contain at least one underscore and one whitespace character");
  const l = [1];
  for (let i = 1; i < o.length; i += 1) {
    const u = o[i - 1], h = o[i];
    u === h ? l[l.length - 1] += 1 : l.push(1);
  }
  return l;
}
function ke(r, e) {
  return ["interpolate", ["linear"], ["get", e], ...r.flatMap((t) => [t.value, t.color])];
}
function Fe(r, e, t = !0) {
  return t ? [
    "interpolate",
    ["linear"],
    ["zoom"],
    0,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.value, o.pointRadius * 0.025])],
    2,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.value, o.pointRadius * 0.05])],
    4,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.value, o.pointRadius * 0.1])],
    8,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.value, o.pointRadius * 0.25])],
    16,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.value, o.pointRadius])]
  ] : ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.value, o.pointRadius])];
}
function go(r, e, t = !0) {
  return t ? [
    "interpolate",
    ["linear"],
    ["zoom"],
    0,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.propertyValue, o.value * 0.025])],
    2,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.propertyValue, o.value * 0.05])],
    4,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.propertyValue, o.value * 0.1])],
    8,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.propertyValue, o.value * 0.25])],
    16,
    ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.propertyValue, o.value])]
  ] : ["interpolate", ["linear"], ["get", e], ...r.flatMap((o) => [o.propertyValue, o.value])];
}
function Oe(r, e) {
  return r.every((t) => t.color[3] === r[0].color[3]) ? r[0].color[3] ? r[0].color[3] / 255 : 1 : [
    "interpolate",
    ["linear"],
    ["get", e],
    ...r.getRawColorStops().flatMap((t) => {
      const o = t.value, a = t.color;
      return [o, a.length === 4 ? a[3] / 255 : 1];
    })
  ];
}
function yo(r, e = 10) {
  return [
    "interpolate",
    ["linear"],
    ["heatmap-density"],
    ...Array.from({ length: e + 1 }, (t, o) => {
      const a = o / e;
      return [a, r.getColorHex(a)];
    }).flat()
  ];
}
async function vo(r, e, t = {}) {
  var a, n;
  if (!e.sourceId && !e.data)
    throw new Error("Creating a polyline layer requires an existing .sourceId or a valid .data property");
  let o = e.data;
  if (typeof o == "string") {
    if (ee(o))
      o = `https://api.maptiler.com/data/${e.data}/features.json?key=${M.apiKey}`;
    else if (((a = o.split(".").pop()) == null ? void 0 : a.toLowerCase().trim()) === "gpx") {
      const i = await (await fetch(o, t)).text();
      o = Ve(i);
    } else if (((n = o.split(".").pop()) == null ? void 0 : n.toLowerCase().trim()) === "kml") {
      const i = await (await fetch(o, t)).text();
      o = $e(i);
    } else {
      const l = Ht(o) ?? co(o);
      l && (o = l);
    }
    if (!o)
      throw new Error("Polyline data was provided as string but is incompatible with valid formats.");
  }
  return Lo(r, {
    ...e,
    data: o
  });
}
function Lo(r, e) {
  if (e.layerId && r.getLayer(e.layerId))
    throw new Error(`A layer already exists with the layer id: ${e.layerId}`);
  const t = e.sourceId ?? te(), o = e.layerId ?? oe(), a = {
    polylineLayerId: o,
    polylineOutlineLayerId: "",
    polylineSourceId: t
  };
  e.data && !r.getSource(t) && r.addSource(t, {
    type: "geojson",
    data: e.data
  });
  const n = e.lineWidth ?? 3, l = e.lineColor ?? he(), i = e.lineOpacity ?? 1, u = e.lineBlur ?? 0, h = e.lineGapWidth ?? 0;
  let c = e.lineDashArray ?? null;
  const s = e.outlineWidth ?? 1, d = e.outlineColor ?? "#FFFFFF", p = e.outlineOpacity ?? 1, m = e.outlineBlur ?? 0;
  if (typeof c == "string" && (c = Je(c)), e.outline === !0) {
    const y = `${o}_outline`;
    a.polylineOutlineLayerId = y, r.addLayer(
      {
        id: y,
        type: "line",
        source: t,
        layout: {
          "line-join": e.lineJoin ?? "round",
          "line-cap": e.lineCap ?? "round"
        },
        minzoom: e.minzoom ?? 0,
        maxzoom: e.maxzoom ?? 23,
        paint: {
          "line-opacity": typeof p == "number" ? p : A(p),
          "line-color": typeof d == "string" ? d : K(d),
          "line-width": mo(n, s),
          "line-blur": typeof m == "number" ? m : A(m)
        }
      },
      e.beforeId
    );
  }
  return r.addLayer(
    {
      id: o,
      type: "line",
      source: t,
      layout: {
        "line-join": e.lineJoin ?? "round",
        "line-cap": e.lineCap ?? "round"
      },
      minzoom: e.minzoom ?? 0,
      maxzoom: e.maxzoom ?? 23,
      paint: {
        "line-opacity": typeof i == "number" ? i : A(i),
        "line-color": typeof l == "string" ? l : K(l),
        "line-width": typeof n == "number" ? n : A(n),
        "line-blur": typeof u == "number" ? u : A(u),
        "line-gap-width": typeof h == "number" ? h : A(h),
        // For some reasons passing "line-dasharray" with the value "undefined"
        // results in no showing the line while it should have the same behavior
        // of not adding the property "line-dasharray" as all.
        // As a workaround, we are inlining the addition of the prop with a conditional
        // which is less readable.
        ...c && { "line-dasharray": c }
      }
    },
    e.beforeId
  ), a;
}
function wo(r, e) {
  if (e.layerId && r.getLayer(e.layerId))
    throw new Error(`A layer already exists with the layer id: ${e.layerId}`);
  const t = e.sourceId ?? te(), o = e.layerId ?? oe(), a = {
    polygonLayerId: o,
    polygonOutlineLayerId: e.outline ? `${o}_outline` : "",
    polygonSourceId: t
  };
  if (e.data && !r.getSource(t)) {
    let y = e.data;
    typeof y == "string" && ee(y) && (y = `https://api.maptiler.com/data/${y}/features.json?key=${M.apiKey}`), r.addSource(t, {
      type: "geojson",
      data: y
    });
  }
  let n = e.outlineDashArray ?? null;
  const l = e.outlineWidth ?? 1, i = e.outlineColor ?? "#FFFFFF", u = e.outlineOpacity ?? 1, h = e.outlineBlur ?? 0, c = e.fillColor ?? he(), s = e.fillOpacity ?? 1, d = e.outlinePosition ?? "center", p = e.pattern ?? null;
  typeof n == "string" && (n = Je(n));
  const m = (y = null) => {
    if (r.addLayer(
      {
        id: o,
        type: "fill",
        source: t,
        minzoom: e.minzoom ?? 0,
        maxzoom: e.maxzoom ?? 23,
        paint: {
          "fill-color": typeof c == "string" ? c : K(c),
          "fill-opacity": typeof s == "number" ? s : A(s),
          // Adding a pattern if provided
          ...y && { "fill-pattern": y }
        }
      },
      e.beforeId
    ), e.outline === !0) {
      let b;
      d === "inside" ? typeof l == "number" ? b = 0.5 * l : b = A(
        l.map(({ zoom: P, value: w }) => ({
          zoom: P,
          value: 0.5 * w
        }))
      ) : d === "outside" ? typeof l == "number" ? b = -0.5 * l : b = A(
        l.map((P) => ({
          zoom: P.zoom,
          value: -0.5 * P.value
        }))
      ) : b = 0, r.addLayer(
        {
          id: a.polygonOutlineLayerId,
          type: "line",
          source: t,
          layout: {
            "line-join": e.outlineJoin ?? "round",
            "line-cap": e.outlineCap ?? "butt"
          },
          minzoom: e.minzoom ?? 0,
          maxzoom: e.maxzoom ?? 23,
          paint: {
            "line-opacity": typeof u == "number" ? u : A(u),
            "line-color": typeof i == "string" ? i : K(i),
            "line-width": typeof l == "number" ? l : A(l),
            "line-blur": typeof h == "number" ? h : A(h),
            "line-offset": b,
            // For some reasons passing "line-dasharray" with the value "undefined"
            // results in no showing the line while it should have the same behavior
            // of not adding the property "line-dasharray" as all.
            // As a workaround, we are inlining the addition of the prop with a conditional
            // which is less readable.
            ...n && {
              "line-dasharray": n
            }
          }
        },
        e.beforeId
      );
    }
  };
  return p ? r.hasImage(p) ? m(p) : r.loadImage(p).then((y) => {
    r.addImage(p, y.data), m(p);
  }).catch((y) => (console.error("Could not load the pattern image.", y.message), m())) : m(), a;
}
function Co(r, e) {
  if (e.layerId && r.getLayer(e.layerId))
    throw new Error(`A layer already exists with the layer id: ${e.layerId}`);
  const t = e.minPointRadius ?? 10, o = e.maxPointRadius ?? 50, a = e.cluster ?? !1, n = 20, l = Array.isArray(e.pointColor) ? e.pointColor : qe.TURBO.scale(10, e.cluster ? 1e4 : 1e3).resample("ease-out-square"), i = l.getBounds(), u = e.sourceId ?? te(), h = e.layerId ?? oe(), c = e.showLabel ?? a, s = e.alignOnViewport ?? !0, d = e.outline ?? !1, p = e.outlineOpacity ?? 1, m = e.outlineWidth ?? 1, y = e.outlineColor ?? "#FFFFFF";
  let b;
  const P = e.zoomCompensation ?? !0, w = e.minzoom ?? 0, _ = e.maxzoom ?? 23;
  typeof e.pointOpacity == "number" ? b = e.pointOpacity : Array.isArray(e.pointOpacity) ? b = A(e.pointOpacity) : e.cluster ? b = Oe(l, "point_count") : e.property ? b = Oe(l, e.property) : b = A([
    { zoom: w, value: 0 },
    { zoom: w + 0.25, value: 1 },
    { zoom: _ - 0.25, value: 1 },
    { zoom: _, value: 0 }
  ]);
  const E = {
    pointLayerId: h,
    clusterLayerId: "",
    labelLayerId: "",
    pointSourceId: u
  };
  if (e.data && !r.getSource(u)) {
    let R = e.data;
    typeof R == "string" && ee(R) && (R = `https://api.maptiler.com/data/${R}/features.json?key=${M.apiKey}`), r.addSource(u, {
      type: "geojson",
      data: R,
      cluster: a
    });
  }
  if (a) {
    E.clusterLayerId = `${h}_cluster`;
    const R = Array.from({ length: n }, (D, V) => {
      const W = i.min + V * (i.max - i.min) / (n - 1);
      return {
        value: W,
        pointRadius: t + (o - t) * (V / (n - 1)) ** 0.5,
        color: l.getColorHex(W)
      };
    });
    r.addLayer(
      {
        id: E.clusterLayerId,
        type: "circle",
        source: u,
        filter: ["has", "point_count"],
        paint: {
          // 'circle-color': options.pointColor ?? colorDrivenByProperty(clusterStyle, "point_count"),
          "circle-color": typeof e.pointColor == "string" ? e.pointColor : ke(R, "point_count"),
          "circle-radius": typeof e.pointRadius == "number" ? e.pointRadius : Array.isArray(e.pointRadius) ? A(e.pointRadius) : Fe(R, "point_count", !1),
          "circle-pitch-alignment": s ? "viewport" : "map",
          "circle-pitch-scale": "map",
          // scale with camera distance regardless of viewport/biewport alignement
          "circle-opacity": b,
          ...d && {
            "circle-stroke-opacity": typeof p == "number" ? p : A(p),
            "circle-stroke-width": typeof m == "number" ? m : A(m),
            "circle-stroke-color": typeof y == "string" ? y : K(y)
          }
        },
        minzoom: w,
        maxzoom: _
      },
      e.beforeId
    ), r.addLayer(
      {
        id: E.pointLayerId,
        type: "circle",
        source: u,
        filter: ["!", ["has", "point_count"]],
        paint: {
          "circle-pitch-alignment": s ? "viewport" : "map",
          "circle-pitch-scale": "map",
          // scale with camera distance regardless of viewport/biewport alignement
          // 'circle-color':  options.pointColor ?? clusterStyle[0].color,
          "circle-color": typeof e.pointColor == "string" ? e.pointColor : l.getColorHex(l.getBounds().min),
          "circle-radius": typeof e.pointRadius == "number" ? e.pointRadius : Array.isArray(e.pointRadius) ? A(e.pointRadius) : R[0].pointRadius * 0.75,
          "circle-opacity": b,
          ...d && {
            "circle-stroke-opacity": typeof p == "number" ? p : A(p),
            "circle-stroke-width": typeof m == "number" ? m : A(m),
            "circle-stroke-color": typeof y == "string" ? y : K(y)
          }
        },
        minzoom: w,
        maxzoom: _
      },
      e.beforeId
    );
  } else {
    let R = typeof e.pointColor == "string" ? e.pointColor : Array.isArray(e.pointColor) ? e.pointColor.getColorHex(e.pointColor.getBounds().min) : he(), D = typeof e.pointRadius == "number" ? P ? A([
      { zoom: 0, value: e.pointRadius * 0.025 },
      { zoom: 2, value: e.pointRadius * 0.05 },
      { zoom: 4, value: e.pointRadius * 0.1 },
      { zoom: 8, value: e.pointRadius * 0.25 },
      { zoom: 16, value: e.pointRadius * 1 }
    ]) : e.pointRadius : Array.isArray(e.pointRadius) ? A(e.pointRadius) : P ? A([
      { zoom: 0, value: t * 0.05 },
      { zoom: 2, value: t * 0.1 },
      { zoom: 4, value: t * 0.2 },
      { zoom: 8, value: t * 0.5 },
      { zoom: 16, value: t * 1 }
    ]) : t;
    if (e.property && Array.isArray(e.pointColor)) {
      const V = Array.from({ length: n }, (W, q) => {
        const J = i.min + q * (i.max - i.min) / (n - 1);
        return {
          value: J,
          pointRadius: typeof e.pointRadius == "number" ? e.pointRadius : t + (o - t) * (q / (n - 1)) ** 0.5,
          color: typeof e.pointColor == "string" ? e.pointColor : l.getColorHex(J)
        };
      });
      R = ke(V, e.property), D = Fe(V, e.property, P);
    }
    r.addLayer(
      {
        id: E.pointLayerId,
        type: "circle",
        source: u,
        layout: {
          // Contrary to labels, we want to see the small one in front. Weirdly "circle-sort-key" works in the opposite direction as "symbol-sort-key".
          "circle-sort-key": e.property ? ["/", 1, ["get", e.property]] : 0
        },
        paint: {
          "circle-pitch-alignment": s ? "viewport" : "map",
          "circle-pitch-scale": "map",
          // scale with camera distance regardless of viewport/biewport alignement
          "circle-color": R,
          "circle-opacity": b,
          "circle-radius": D,
          ...d && {
            "circle-stroke-opacity": typeof p == "number" ? p : A(p),
            "circle-stroke-width": typeof m == "number" ? m : A(m),
            "circle-stroke-color": typeof y == "string" ? y : K(y)
          }
        },
        minzoom: w,
        maxzoom: _
      },
      e.beforeId
    );
  }
  if (c !== !1 && (e.cluster || e.property)) {
    E.labelLayerId = `${h}_label`;
    const R = e.labelColor ?? "#fff", D = e.labelSize ?? 12;
    r.addLayer(
      {
        id: E.labelLayerId,
        type: "symbol",
        source: u,
        filter: ["has", e.cluster ? "point_count" : e.property],
        layout: {
          "text-field": e.cluster ? "{point_count_abbreviated}" : `{${e.property}}`,
          "text-font": ["Noto Sans Regular"],
          "text-size": D,
          "text-pitch-alignment": s ? "viewport" : "map",
          "symbol-sort-key": ["/", 1, ["get", e.cluster ? "point_count" : e.property]]
          // so that the largest value goes on top
        },
        paint: {
          "text-color": R,
          "text-opacity": b
        },
        minzoom: w,
        maxzoom: _
      },
      e.beforeId
    );
  }
  return E;
}
function bo(r, e) {
  if (e.layerId && r.getLayer(e.layerId))
    throw new Error(`A layer already exists with the layer id: ${e.layerId}`);
  const t = e.sourceId ?? te(), o = e.layerId ?? oe(), a = e.minzoom ?? 0, n = e.maxzoom ?? 23, l = e.zoomCompensation ?? !0, i = e.opacity ?? [
    { zoom: a, value: 0 },
    { zoom: a + 0.25, value: 1 },
    { zoom: n - 0.25, value: 1 },
    { zoom: n, value: 0 }
  ];
  let u = Array.isArray(e.colorRamp) ? e.colorRamp : qe.TURBO.transparentStart();
  const h = u.getBounds();
  (h.min !== 0 || h.max !== 1) && (u = u.scale(0, 1)), u.hasTransparentStart() || (u = u.transparentStart());
  const c = e.intensity ?? [
    { zoom: 0, value: 0.01 },
    { zoom: 4, value: 0.2 },
    { zoom: 16, value: 1 }
  ], s = e.property ?? null, d = e.weight ?? 1;
  let p = 1;
  s ? typeof d == "number" ? (p = d, typeof e.weight == "number" && console.warn(
    "The option `.property` is ignored when `.propertyValueWeights` is not of type `PropertyValueWeights`"
  )) : Array.isArray(d) ? p = fo(d, s) : console.warn(
    "The option `.property` is ignored when `.propertyValueWeights` is not of type `PropertyValueWeights`"
  ) : typeof d == "number" ? p = d : Array.isArray(d) && console.warn("The options `.propertyValueWeights` can only be used when `.property` is provided.");
  const m = [
    { zoom: 0, value: 50 * 0.025 },
    { zoom: 2, value: 50 * 0.05 },
    { zoom: 4, value: 50 * 0.1 },
    { zoom: 8, value: 50 * 0.25 },
    { zoom: 16, value: 50 }
  ], y = e.radius ?? (l ? m : 10);
  let b = 1;
  typeof y == "number" ? b = y : Array.isArray(y) && "zoom" in y[0] ? b = A(y) : s && Array.isArray(y) && "propertyValue" in y[0] ? b = go(y, s, l) : !s && Array.isArray(y) && "propertyValue" in y[0] ? (b = A(m), console.warn("The option `.radius` can only be property-driven if the option `.property` is provided.")) : b = A(m);
  const P = {
    heatmapLayerId: o,
    heatmapSourceId: t
  };
  if (e.data && !r.getSource(t)) {
    let w = e.data;
    typeof w == "string" && ee(w) && (w = `https://api.maptiler.com/data/${w}/features.json?key=${M.apiKey}`), r.addSource(t, {
      type: "geojson",
      data: w
    });
  }
  return r.addLayer(
    {
      id: o,
      type: "heatmap",
      source: t,
      minzoom: a,
      maxzoom: n,
      paint: {
        "heatmap-weight": p,
        "heatmap-intensity": typeof c == "number" ? c : A(c),
        "heatmap-color": yo(u),
        "heatmap-radius": b,
        "heatmap-opacity": typeof i == "number" ? i : A(i)
      }
    },
    e.beforeId
  ), P;
}
const qo = {
  addPolyline: vo,
  addPolygon: wo,
  addPoint: Co,
  addHeatmap: bo,
  takeScreenshot: po
};
Ut();
function Jo() {
  return Pt.version;
}
const Xo = f.Map, Qo = f.Marker, er = f.Popup, tr = f.Style, or = f.CanvasSource, rr = f.GeoJSONSource, ar = f.ImageSource, nr = f.RasterTileSource, lr = f.RasterDEMTileSource, ir = f.VectorTileSource, sr = f.VideoSource, cr = f.NavigationControl, ur = f.GeolocateControl, dr = f.AttributionControl, pr = f.LogoControl, hr = f.ScaleControl, mr = f.FullscreenControl, fr = f.TerrainControl, gr = f.BoxZoomHandler, yr = f.ScrollZoomHandler, vr = f.CooperativeGesturesHandler, Lr = f.KeyboardHandler, wr = f.TwoFingersTouchPitchHandler, Cr = f.MapWheelEvent, br = f.MapTouchEvent, Sr = f.MapMouseEvent, Ar = f.config, Tr = f.getVersion, {
  setRTLTextPlugin: Rr,
  getRTLTextPluginStatus: _r,
  LngLat: Er,
  LngLatBounds: xr,
  MercatorCoordinate: Ir,
  Evented: Mr,
  AJAXError: Pr,
  prewarm: kr,
  clearPrewarmedResources: Fr,
  Hash: Or,
  Point: Br,
  EdgeInsets: Nr,
  DragRotateHandler: zr,
  DragPanHandler: Dr,
  TwoFingersTouchZoomRotateHandler: Ur,
  DoubleClickZoomHandler: Gr,
  TwoFingersTouchZoomHandler: Hr,
  TwoFingersTouchRotateHandler: Vr,
  getWorkerCount: $r,
  setWorkerCount: jr,
  getMaxParallelImageRequests: Kr,
  setMaxParallelImageRequests: Zr,
  getWorkerUrl: Wr,
  setWorkerUrl: Yr,
  addSourceType: qr,
  importScriptInWorkers: Jr,
  addProtocol: Xr,
  removeProtocol: Qr
} = f;
export {
  Pr as AJAXError,
  Uo as AttributionControl,
  dr as AttributionControlMLGL,
  Ho as BoxZoomHandler,
  gr as BoxZoomHandlerMLGL,
  ko as CanvasSource,
  or as CanvasSourceMLGL,
  g as ColorRamp,
  qe as ColorRampCollection,
  $o as CooperativeGesturesHandler,
  vr as CooperativeGesturesHandlerMLGL,
  Gr as DoubleClickZoomHandler,
  Dr as DragPanHandler,
  zr as DragRotateHandler,
  Nr as EdgeInsets,
  Mr as Evented,
  Yt as FullscreenControl,
  mr as FullscreenControlMLGL,
  Fo as GeoJSONSource,
  rr as GeoJSONSourceMLGL,
  Kt as GeolocateControl,
  ur as GeolocateControlMLGL,
  eo as GeolocationType,
  Or as Hash,
  Oo as ImageSource,
  ar as ImageSourceMLGL,
  jo as KeyboardHandler,
  Lr as KeyboardHandlerMLGL,
  C as Language,
  Er as LngLat,
  xr as LngLatBounds,
  Zt as LogoControl,
  pr as LogoControlMLGL,
  to as Map,
  Xo as MapMLGL,
  Yo as MapMouseEvent,
  Sr as MapMouseEventMLGL,
  oa as MapStyle,
  ra as MapStyleVariant,
  Wo as MapTouchEvent,
  br as MapTouchEventMLGL,
  Zo as MapWheelEvent,
  Cr as MapWheelEventMLGL,
  Qt as MaptilerGeolocateControl,
  be as MaptilerLogoControl,
  Jt as MaptilerNavigationControl,
  qt as MaptilerTerrainControl,
  xo as Marker,
  Qo as MarkerMLGL,
  Ir as MercatorCoordinate,
  cr as NavigationControMLGL,
  jt as NavigationControl,
  Br as Point,
  Mo as Popup,
  er as PopupMLGL,
  No as RasterDEMTileSource,
  lr as RasterDEMTileSourceMLGL,
  Bo as RasterTileSource,
  nr as RasterTileSourceMLGL,
  aa as ReferenceMapStyle,
  Wt as ScaleControl,
  hr as ScaleControlMLGL,
  Vo as ScrollZoomHandler,
  yr as ScrollZoomHandlerMLGL,
  kt as SdkConfig,
  na as ServiceError,
  Po as Style,
  tr as StyleMLGL,
  fr as TerrainControMLGL,
  Go as TerrainControl,
  Ko as TwoFingersTouchPitchHandler,
  wr as TwoFingersTouchPitchHandlerMLGL,
  Vr as TwoFingersTouchRotateHandler,
  Hr as TwoFingersTouchZoomHandler,
  Ur as TwoFingersTouchZoomRotateHandler,
  zo as VectorTileSource,
  ir as VectorTileSourceMLGL,
  Do as VideoSource,
  sr as VideoSourceMLGL,
  Xr as addProtocol,
  qr as addSourceType,
  la as areSameLanguages,
  ia as bufferToPixelDataBrowser,
  sa as circumferenceAtLatitude,
  Fr as clearPrewarmedResources,
  M as config,
  Ar as configMLGL,
  ca as coordinates,
  ua as data,
  da as elevation,
  pa as expandMapStyle,
  ha as geocoding,
  ma as geolocation,
  fa as getAutoLanguage,
  ye as getBrowserLanguage,
  ga as getBufferToPixelDataParser,
  ya as getLanguageInfoFromCode,
  va as getLanguageInfoFromFlag,
  La as getLanguageInfoFromKey,
  Tr as getMapLibreVersion,
  Kr as getMaxParallelImageRequests,
  _r as getRTLTextPluginStatus,
  wa as getTileCache,
  Jo as getVersion,
  $r as getWorkerCount,
  Wr as getWorkerUrl,
  Ve as gpx,
  co as gpxOrKml,
  He as hasChildNodeWithName,
  qo as helpers,
  Jr as importScriptInWorkers,
  Ca as isLanguageInfo,
  $e as kml,
  ba as mapStylePresetList,
  Sa as math,
  Aa as misc,
  kr as prewarm,
  Qr as removeProtocol,
  Zr as setMaxParallelImageRequests,
  Rr as setRTLTextPlugin,
  jr as setWorkerCount,
  Yr as setWorkerUrl,
  Ta as staticMaps,
  de as str2xml,
  Ra as styleToStyle,
  _a as toLanguageInfo,
  Re as xml2str
};
